<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Linkage Disequilibrium Shrinkage Estimation for Polyploids • ldsep</title>
<!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="favicon-16x16.png">
<link rel="icon" type="image/png" sizes="32x32" href="favicon-32x32.png">
<link rel="apple-touch-icon" type="image/png" sizes="180x180" href="apple-touch-icon.png">
<link rel="apple-touch-icon" type="image/png" sizes="120x120" href="apple-touch-icon-120x120.png">
<link rel="apple-touch-icon" type="image/png" sizes="76x76" href="apple-touch-icon-76x76.png">
<link rel="apple-touch-icon" type="image/png" sizes="60x60" href="apple-touch-icon-60x60.png">
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="bootstrap-toc.css">
<script src="bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="pkgdown.css" rel="stylesheet">
<script src="pkgdown.js"></script><meta property="og:title" content="Linkage Disequilibrium Shrinkage Estimation for Polyploids">
<meta property="og:description" content="Estimate haplotypic or composite pairwise linkage disequilibrium
    (LD) in polyploids, using either genotypes or genotype likelihoods. 
    Support is provided to estimate the popular measures of LD: the LD 
    coefficient D, the standardized LD coefficient D', and the Pearson 
    correlation coefficient r. All estimates are returned with corresponding 
    standard errors. These estimates and standard errors can then be used
    for shrinkage estimation. The main functions are ldfast(), ldest(), mldest(),
    sldest(), plot.lddf(), format_lddf(), and ldshrink(). Details of the methods
    are available in Gerard (2021a) &lt;doi:10.1111/1755-0998.13349&gt;
    and Gerard (2021b) &lt;doi:10.1101/2021.02.08.430270&gt;.">
<meta property="og:image" content="https://dcgerard.github.io/ldsep/logo.png">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    <div class="container template-home">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="index.html">ldsep</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">2.1.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="index.html">
    <span class="fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="articles/fast.html">Fast LD Estimation</a>
    </li>
    <li>
      <a href="articles/vcf.html">LD Estimation Pipeline from VCF</a>
    </li>
  </ul>
</li>
<li>
  <a href="news/index.html">Changelog</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/dcgerard/ldsep/">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="contents col-md-9">
<div id="ldsep-linkage-disequilibrium-shrinkage-estimation-for-polyploids-" class="section level1">
<div class="page-header"><h1 class="hasAnchor">
<a href="#ldsep-linkage-disequilibrium-shrinkage-estimation-for-polyploids-" class="anchor"></a>ldsep: Linkage Disequilibrium Shrinkage Estimation for Polyploids <a href="https://dcgerard.github.io/ldsep/"><img src="reference/figures/logo.png" align="right" height="138"></a>
</h1></div>
<!-- badges: start -->

<p>Estimate haplotypic or composite pairwise linkage disequilibrium (LD) in polyploids, using either genotypes or genotype likelihoods. Support is provided to estimate the popular measures of LD: the LD coefficient <em>D</em>, the standardized LD coefficient <em>D</em>′, and the Pearson correlation coefficient <em>r</em>. All estimates are returned with corresponding standard errors. These estimates and standard errors can then be used for shrinkage estimation. The methods are described in Gerard (2021a) and Gerard (2021b).</p>
<p>The main functions are:</p>
<ul>
<li>
<code><a href="reference/ldfast.html">ldfast()</a></code>: Fast, moment-based approach to estimate pairwise LD in the presence of genotype uncertainty.</li>
<li>
<code><a href="reference/ldest.html">ldest()</a></code>: Estimates pairwise LD via maximum likelihood.</li>
<li>
<code><a href="reference/mldest.html">mldest()</a></code>: Iteratively apply <code><a href="reference/ldest.html">ldest()</a></code> across many pairs of SNPs.</li>
<li>
<code><a href="reference/sldest.html">sldest()</a></code>: Iteratively apply <code><a href="reference/ldest.html">ldest()</a></code> along a sliding window of fixed length.</li>
<li>
<code><a href="reference/plot.lddf.html">plot.lddf()</a></code>: Plot method for the output of <code><a href="reference/mldest.html">mldest()</a></code> and <code><a href="reference/sldest.html">sldest()</a></code>.</li>
<li>
<code><a href="reference/format_lddf.html">format_lddf()</a></code>: Format the output of <code><a href="reference/mldest.html">mldest()</a></code> and <code><a href="reference/sldest.html">sldest()</a></code> into a matrix.</li>
<li>
<code><a href="reference/ldshrink.html">ldshrink()</a></code>: Shrink correlation estimates using adaptive shrinkage (Stephens, 2017; Dey and Stephens, 2018).</li>
</ul>
<div id="installation" class="section level2">
<h2 class="hasAnchor">
<a href="#installation" class="anchor"></a>Installation</h2>
<p>You can install the released version of ldsep from <a href="https://cran.r-project.org/package=ldsep">CRAN</a> with:</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html">install.packages</a></span><span class="op">(</span><span class="st">"ldsep"</span><span class="op">)</span></code></pre></div>
<p>And the development version from <a href="https://github.com/dcgerard/ldsep">GitHub</a> with:</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># install.packages("devtools")</span>
<span class="fu">devtools</span><span class="fu">::</span><span class="fu"><a href="https://devtools.r-lib.org//reference/remote-reexports.html">install_github</a></span><span class="op">(</span><span class="st">"dcgerard/ldsep"</span><span class="op">)</span></code></pre></div>
</div>
<div id="citation" class="section level2">
<h2 class="hasAnchor">
<a href="#citation" class="anchor"></a>Citation</h2>
<p>To cite <code>ldsep</code> in publications use:</p>
<blockquote>
<p>Gerard, David (2021). “Pairwise Linkage Disequilibrium Estimation for Polyploids.” <em>Molecular Ecology Resources</em>, <em>21</em>(4), 1230–1242. <a href="https://doi.org/10.1111/1755-0998.13349">doi:10.1111/1755-0998.13349</a>.</p>
</blockquote>
<p>A BibTeX entry for LaTeX users is</p>
<div class="sourceCode" id="cb3"><pre class="sourceCode tex"><code class="sourceCode latex"><span id="cb3-1"><a href="#cb3-1" aria-hidden="true" tabindex="-1"></a>@Article{,</span>
<span id="cb3-2"><a href="#cb3-2" aria-hidden="true" tabindex="-1"></a>  title = {Pairwise Linkage Disequilibrium Estimation for Polyploids},</span>
<span id="cb3-3"><a href="#cb3-3" aria-hidden="true" tabindex="-1"></a>  author = {David Gerard},</span>
<span id="cb3-4"><a href="#cb3-4" aria-hidden="true" tabindex="-1"></a>  journal = {Molecular Ecology Resources},</span>
<span id="cb3-5"><a href="#cb3-5" aria-hidden="true" tabindex="-1"></a>  year = {2021},</span>
<span id="cb3-6"><a href="#cb3-6" aria-hidden="true" tabindex="-1"></a>  doi = {10.1111/1755-0998.13349},</span>
<span id="cb3-7"><a href="#cb3-7" aria-hidden="true" tabindex="-1"></a>  volume = {21},</span>
<span id="cb3-8"><a href="#cb3-8" aria-hidden="true" tabindex="-1"></a>  number = {4},</span>
<span id="cb3-9"><a href="#cb3-9" aria-hidden="true" tabindex="-1"></a>  pages = {1230--1242},</span>
<span id="cb3-10"><a href="#cb3-10" aria-hidden="true" tabindex="-1"></a>}</span></code></pre></div>
<p>If you use <code><a href="reference/ldfast.html">ldfast()</a></code>, please cite:</p>
<blockquote>
<p>Gerard, David (2021). “Scalable Bias-corrected Linkage Disequilibrium Estimation Under Genotype Uncertainty.” <em>bioRxiv</em>. <a href="https://doi.org/10.1101/2021.02.08.430270">doi:10.1101/2021.02.08.430270</a>.</p>
</blockquote>
<p>A BibTeX entry for LaTeX users is</p>
<div class="sourceCode" id="cb4"><pre class="sourceCode tex"><code class="sourceCode latex"><span id="cb4-1"><a href="#cb4-1" aria-hidden="true" tabindex="-1"></a>@Article{,</span>
<span id="cb4-2"><a href="#cb4-2" aria-hidden="true" tabindex="-1"></a>  title = {Scalable Bias-corrected Linkage Disequilibrium Estimation Under Genotype Uncertainty},</span>
<span id="cb4-3"><a href="#cb4-3" aria-hidden="true" tabindex="-1"></a>  author = {David Gerard},</span>
<span id="cb4-4"><a href="#cb4-4" aria-hidden="true" tabindex="-1"></a>  journal = {bioRxiv},</span>
<span id="cb4-5"><a href="#cb4-5" aria-hidden="true" tabindex="-1"></a>  year = {2021},</span>
<span id="cb4-6"><a href="#cb4-6" aria-hidden="true" tabindex="-1"></a>  doi = {10.1101/2021.02.08.430270},</span>
<span id="cb4-7"><a href="#cb4-7" aria-hidden="true" tabindex="-1"></a>}</span></code></pre></div>
</div>
<div id="code-of-conduct" class="section level2">
<h2 class="hasAnchor">
<a href="#code-of-conduct" class="anchor"></a>Code of Conduct</h2>
<p>Please note that the ldsep project is released with a <a href="https://contributor-covenant.org/version/2/0/CODE_OF_CONDUCT.html">Contributor Code of Conduct</a>. By contributing to this project, you agree to abide by its terms.</p>
</div>
<div id="references" class="section level2">
<h2 class="hasAnchor">
<a href="#references" class="anchor"></a>References</h2>
<ul>
<li><p>Dey, Kushal K., and Matthew Stephens (2018). “CorShrink: Empirical Bayes shrinkage estimation of correlations, with applications.” <em>bioRxiv</em>. <a href="https://doi.org/10.1101/368316">doi:10.1101/368316</a></p></li>
<li><p>Gerard, David (2021a). “Pairwise Linkage Disequilibrium Estimation for Polyploids.” <em>Molecular Ecology Resources</em>, 21(4), 1230–1242. <a href="https://doi.org/10.1111/1755-0998.13349">doi:10.1111/1755-0998.13349</a>.</p></li>
<li><p>Gerard, David (2021b). “Scalable Bias-corrected Linkage Disequilibrium Estimation Under Genotype Uncertainty.” <em>bioRxiv</em>. <a href="https://doi.org/10.1101/2021.02.08.430270">doi:10.1101/2021.02.08.430270</a></p></li>
<li><p>Stephens, Matthew (2017). “False discovery rates: a new deal.” <em>Biostatistics</em> 18(2), 275–294. <a href="https://doi.org/10.1093/biostatistics/kxw041">doi:10.1093/biostatistics/kxw041</a></p></li>
</ul>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <div class="links">
<h2>Links</h2>
<ul class="list-unstyled">
<li>Download from CRAN at <br><a href="https://cloud.r-project.org/package=ldsep">https://​cloud.r-project.org/​package=ldsep</a>
</li>
<li>Browse source code at <br><a href="https://github.com/dcgerard/ldsep/">https://​github.com/​dcgerard/​ldsep/​</a>
</li>
<li>Report a bug at <br><a href="https://github.com/dcgerard/ldsep/issues">https://​github.com/​dcgerard/​ldsep/​issues</a>
</li>
</ul>
</div>
<div class="license">
<h2>License</h2>
<ul class="list-unstyled">
<li><a href="LICENSE.html">Full license</a></li>
<li><small><a href="https://www.r-project.org/Licenses/GPL-3">GPL-3</a></small></li>
</ul>
</div>
<div class="community">
<h2>Community</h2>
<ul class="list-unstyled">
<li><a href="CODE_OF_CONDUCT.html">Code of conduct</a></li>
</ul>
</div>
<div class="citation">
<h2>Citation</h2>
<ul class="list-unstyled">
<li><a href="authors.html">Citing ldsep</a></li>
</ul>
</div>
<div class="developers">
<h2>Developers</h2>
<ul class="list-unstyled">
<li>David Gerard <br><small class="roles"> Author, maintainer </small> <a href="https://orcid.org/0000-0001-9450-5023" target="orcid.widget" aria-label="ORCID"><span class="fab fa-orcid orcid" aria-hidden="true"></span></a> </li>
</ul>
</div>

  <div class="dev-status">
<h2>Dev status</h2>
<ul class="list-unstyled">
<li><a href="https://lifecycle.r-lib.org/articles/stages.html"><img src="https://img.shields.io/badge/lifecycle-stable-brightgreen.svg" alt="Lifecycle: stable"></a></li>
<li><a href="https://github.com/dcgerard/ldsep/actions"><img src="https://github.com/dcgerard/ldsep/workflows/R-CMD-check/badge.svg" alt="R build status"></a></li>
<li><a href="https://codecov.io/gh/dcgerard/ldsep"><img src="https://codecov.io/gh/dcgerard/ldsep/branch/master/graph/badge.svg?token=J3ZRB8PAHY" alt="codecov"></a></li>
<li><a href="https://CRAN.R-project.org/package=ldsep"><img src="https://www.r-pkg.org/badges/version/ldsep" alt="CRAN status"></a></li>
<li><a href="https://cran.r-project.org/package=ldsep"><img src="http://cranlogs.r-pkg.org/badges/grand-total/ldsep"></a></li>
<li><a href="https://www.gnu.org/licenses/gpl-3.0"><img src="https://img.shields.io/badge/License-GPL%20v3-blue.svg" alt="License: GPL v3"></a></li>
</ul>
</div>
</div>
</div>


      <footer><div class="copyright">
  <p>Developed by David Gerard.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.6.1.</p>
</div>

      </footer>
</div>

  


  </body>
</html>
