<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<title>Linkage Disequilibrium Shrinkage Estimation for Polyploids • ldsep</title>
<!-- favicons --><link rel="icon" type="image/png" sizes="96x96" href="favicon-96x96.png">
<link rel="icon" type="”image/svg+xml”" href="favicon.svg">
<link rel="apple-touch-icon" sizes="180x180" href="apple-touch-icon.png">
<link rel="icon" sizes="any" href="favicon.ico">
<link rel="manifest" href="site.webmanifest">
<script src="deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet">
<link href="deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet">
<script src="deps/headroom-0.11.0/headroom.min.js"></script><script src="deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="deps/search-1.0.0/fuse.min.js"></script><script src="deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="pkgdown.js"></script><meta property="og:title" content="Linkage Disequilibrium Shrinkage Estimation for Polyploids">
<meta name="description" content="Estimate haplotypic or composite pairwise linkage disequilibrium (LD) in polyploids, using either genotypes or genotype likelihoods. Support is provided to estimate the popular measures of LD: the LD coefficient D, the standardized LD coefficient D', and the Pearson correlation coefficient r. All estimates are returned with corresponding standard errors. These estimates and standard errors can then be used for shrinkage estimation. The main functions are ldfast(), ldest(), mldest(), sldest(), plot.lddf(), format_lddf(), and ldshrink(). Details of the methods are available in Gerard (2021a) &lt;doi:10.1111/1755-0998.13349&gt; and Gerard (2021b) &lt;doi:10.1038/s41437-021-00462-5&gt;.">
<meta property="og:description" content="Estimate haplotypic or composite pairwise linkage disequilibrium (LD) in polyploids, using either genotypes or genotype likelihoods. Support is provided to estimate the popular measures of LD: the LD coefficient D, the standardized LD coefficient D', and the Pearson correlation coefficient r. All estimates are returned with corresponding standard errors. These estimates and standard errors can then be used for shrinkage estimation. The main functions are ldfast(), ldest(), mldest(), sldest(), plot.lddf(), format_lddf(), and ldshrink(). Details of the methods are available in Gerard (2021a) &lt;doi:10.1111/1755-0998.13349&gt; and Gerard (2021b) &lt;doi:10.1038/s41437-021-00462-5&gt;.">
<meta property="og:image" content="https://dcgerard.github.io/ldsep/logo.png">
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="index.html">ldsep</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">2.1.6</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item"><a class="nav-link" href="reference/index.html">Reference</a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles">
<li><a class="dropdown-item" href="articles/fast.html">Fast LD Estimation</a></li>
    <li><a class="dropdown-item" href="articles/vcf.html">LD Estimation Pipeline from VCF</a></li>
  </ul>
</li>
<li class="nav-item"><a class="nav-link" href="news/index.html">Changelog</a></li>
      </ul>
<ul class="navbar-nav">
<li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="search.json">
</form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/dcgerard/ldsep/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
      </ul>
</div>


  </div>
</nav><div class="container template-home">
<div class="row">
  <main id="main" class="col-md-9"><div class="section level1">
<div class="page-header">
<img src="logo.png" class="logo" alt=""><h1 id="ldsep-linkage-disequilibrium-shrinkage-estimation-for-polyploids-">ldsep: Linkage Disequilibrium Shrinkage Estimation for Polyploids 
<a class="anchor" aria-label="anchor" href="#ldsep-linkage-disequilibrium-shrinkage-estimation-for-polyploids-"></a>
</h1>
</div>
<!-- badges: start -->

<p>Estimate haplotypic or composite pairwise linkage disequilibrium (LD) in polyploids, using either genotypes or genotype likelihoods. Support is provided to estimate the popular measures of LD: the LD coefficient <math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>D</mi><annotation encoding="application/x-tex">D</annotation></semantics></math>, the standardized LD coefficient <math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msup><mi>D</mi><mo>′</mo></msup><annotation encoding="application/x-tex">D'</annotation></semantics></math>, and the Pearson correlation coefficient <math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>r</mi><annotation encoding="application/x-tex">r</annotation></semantics></math>. All estimates are returned with corresponding standard errors. These estimates and standard errors can then be used for shrinkage estimation. The methods are described in Gerard (2021a) and Gerard (2021b).</p>
<p>The main functions are:</p>
<ul>
<li>
<code><a href="reference/ldfast.html">ldfast()</a></code>: Fast, moment-based approach to estimate pairwise LD in the presence of genotype uncertainty.</li>
<li>
<code><a href="reference/ldest.html">ldest()</a></code>: Estimates pairwise LD via maximum likelihood.</li>
<li>
<code><a href="reference/mldest.html">mldest()</a></code>: Iteratively apply <code><a href="reference/ldest.html">ldest()</a></code> across many pairs of SNPs.</li>
<li>
<code><a href="reference/sldest.html">sldest()</a></code>: Iteratively apply <code><a href="reference/ldest.html">ldest()</a></code> along a sliding window of fixed length.</li>
<li>
<code><a href="reference/plot.lddf.html">plot.lddf()</a></code>: Plot method for the output of <code><a href="reference/mldest.html">mldest()</a></code> and <code><a href="reference/sldest.html">sldest()</a></code>.</li>
<li>
<code><a href="reference/format_lddf.html">format_lddf()</a></code>: Format the output of <code><a href="reference/mldest.html">mldest()</a></code> and <code><a href="reference/sldest.html">sldest()</a></code> into a matrix.</li>
<li>
<code><a href="reference/ldshrink.html">ldshrink()</a></code>: Shrink correlation estimates using adaptive shrinkage (Stephens, 2017; Dey and Stephens, 2018).</li>
</ul>
<div class="section level2">
<h2 id="installation">Installation<a class="anchor" aria-label="anchor" href="#installation"></a>
</h2>
<p>You can install the released version of ldsep from <a href="https://cran.r-project.org/package=ldsep" class="external-link">CRAN</a> with:</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html" class="external-link">install.packages</a></span><span class="op">(</span><span class="st">"ldsep"</span><span class="op">)</span></span></code></pre></div>
<p>And the development version from <a href="https://github.com/dcgerard/ldsep" class="external-link">GitHub</a> with:</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># install.packages("devtools")</span></span>
<span><span class="fu">devtools</span><span class="fu">::</span><span class="fu"><a href="https://remotes.r-lib.org/reference/install_github.html" class="external-link">install_github</a></span><span class="op">(</span><span class="st">"dcgerard/ldsep"</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="citation">Citation<a class="anchor" aria-label="anchor" href="#citation"></a>
</h2>
<p>To cite <code>ldsep</code> in publications use:</p>
<blockquote>
<p>Gerard, David (2021). “Pairwise Linkage Disequilibrium Estimation for Polyploids.” <em>Molecular Ecology Resources</em>, <em>21</em>(4), 1230–1242. <a href="https://doi.org/10.1111/1755-0998.13349" class="external-link">doi:10.1111/1755-0998.13349</a>.</p>
</blockquote>
<p>A BibTeX entry for LaTeX users is</p>
<div class="sourceCode" id="cb3"><pre class="sourceCode tex"><code class="sourceCode latex"><span id="cb3-1"><a href="#cb3-1" tabindex="-1"></a>@Article{,</span>
<span id="cb3-2"><a href="#cb3-2" tabindex="-1"></a>  title = {Pairwise Linkage Disequilibrium Estimation for Polyploids},</span>
<span id="cb3-3"><a href="#cb3-3" tabindex="-1"></a>  author = {David Gerard},</span>
<span id="cb3-4"><a href="#cb3-4" tabindex="-1"></a>  journal = {Molecular Ecology Resources},</span>
<span id="cb3-5"><a href="#cb3-5" tabindex="-1"></a>  year = {2021},</span>
<span id="cb3-6"><a href="#cb3-6" tabindex="-1"></a>  doi = {10.1111/1755-0998.13349},</span>
<span id="cb3-7"><a href="#cb3-7" tabindex="-1"></a>  volume = {21},</span>
<span id="cb3-8"><a href="#cb3-8" tabindex="-1"></a>  number = {4},</span>
<span id="cb3-9"><a href="#cb3-9" tabindex="-1"></a>  pages = {1230--1242},</span>
<span id="cb3-10"><a href="#cb3-10" tabindex="-1"></a>}</span></code></pre></div>
<p>If you use <code><a href="reference/ldfast.html">ldfast()</a></code>, please cite:</p>
<blockquote>
<p>Gerard, David (2021). “Scalable Bias-corrected Linkage Disequilibrium Estimation Under Genotype Uncertainty.” <em>Heredity</em>, <em>127</em>(4), 357–362. <a href="https://doi.org/10.1038/s41437-021-00462-5" class="external-link">doi:10.1038/s41437-021-00462-5</a>.</p>
</blockquote>
<p>A BibTeX entry for LaTeX users is</p>
<div class="sourceCode" id="cb4"><pre class="sourceCode tex"><code class="sourceCode latex"><span id="cb4-1"><a href="#cb4-1" tabindex="-1"></a>@Article{,</span>
<span id="cb4-2"><a href="#cb4-2" tabindex="-1"></a>  title = {Scalable Bias-corrected Linkage Disequilibrium Estimation Under Genotype Uncertainty},</span>
<span id="cb4-3"><a href="#cb4-3" tabindex="-1"></a>  author = {David Gerard},</span>
<span id="cb4-4"><a href="#cb4-4" tabindex="-1"></a>  journal = {Heredity},</span>
<span id="cb4-5"><a href="#cb4-5" tabindex="-1"></a>  year = {2021},</span>
<span id="cb4-6"><a href="#cb4-6" tabindex="-1"></a>  volume = {127},</span>
<span id="cb4-7"><a href="#cb4-7" tabindex="-1"></a>  number = {4},</span>
<span id="cb4-8"><a href="#cb4-8" tabindex="-1"></a>  pages = {357--362},</span>
<span id="cb4-9"><a href="#cb4-9" tabindex="-1"></a>  doi = {10.1038/s41437-021-00462-5},</span>
<span id="cb4-10"><a href="#cb4-10" tabindex="-1"></a>}</span></code></pre></div>
</div>
<div class="section level2">
<h2 id="code-of-conduct">Code of Conduct<a class="anchor" aria-label="anchor" href="#code-of-conduct"></a>
</h2>
<p>Please note that the ldsep project is released with a <a href="https://contributor-covenant.org/version/2/0/CODE_OF_CONDUCT.html" class="external-link">Contributor Code of Conduct</a>. By contributing to this project, you agree to abide by its terms.</p>
</div>
<div class="section level2">
<h2 id="references">References<a class="anchor" aria-label="anchor" href="#references"></a>
</h2>
<ul>
<li><p>Dey, Kushal K., and Matthew Stephens (2018). “CorShrink: Empirical Bayes shrinkage estimation of correlations, with applications.” <em>bioRxiv</em>. <a href="https://doi.org/10.1101/368316" class="external-link">doi:10.1101/368316</a></p></li>
<li><p>Gerard, David (2021a). “Pairwise Linkage Disequilibrium Estimation for Polyploids.” <em>Molecular Ecology Resources</em>, <em>21</em>(4), 1230–1242. <a href="https://doi.org/10.1111/1755-0998.13349" class="external-link">doi:10.1111/1755-0998.13349</a>.</p></li>
<li><p>Gerard, David (2021b). “Scalable Bias-corrected Linkage Disequilibrium Estimation Under Genotype Uncertainty.” <em>Heredity</em>, <em>127</em>(4), 357–362. <a href="https://doi.org/10.1038/s41437-021-00462-5" class="external-link">doi:10.1038/s41437-021-00462-5</a>.</p></li>
<li><p>Stephens, Matthew (2017). “False discovery rates: a new deal.” <em>Biostatistics</em> 18(2), 275–294. <a href="https://doi.org/10.1093/biostatistics/kxw041" class="external-link">doi:10.1093/biostatistics/kxw041</a></p></li>
</ul>
</div>
</div>
  </main><aside class="col-md-3"><div class="links">
<h2 data-toc-skip>Links</h2>
<ul class="list-unstyled">
<li><a href="https://cloud.r-project.org/package=ldsep" class="external-link">View on CRAN</a></li>
<li><a href="https://github.com/dcgerard/ldsep/" class="external-link">Browse source code</a></li>
<li><a href="https://github.com/dcgerard/ldsep/issues" class="external-link">Report a bug</a></li>
</ul>
</div>

<div class="license">
<h2 data-toc-skip>License</h2>
<ul class="list-unstyled">
<li><a href="LICENSE.html">Full license</a></li>
<li><small>GPL (&gt;= 3)</small></li>
</ul>
</div>

<div class="community">
<h2 data-toc-skip>Community</h2>
<ul class="list-unstyled">
<li><a href="CODE_OF_CONDUCT.html">Code of conduct</a></li>
</ul>
</div>

<div class="citation">
<h2 data-toc-skip>Citation</h2>
<ul class="list-unstyled">
<li><a href="authors.html#citation">Citing ldsep</a></li>
</ul>
</div>

<div class="developers">
<h2 data-toc-skip>Developers</h2>
<ul class="list-unstyled">
<li>David Gerard <br><small class="roles"> Author, maintainer </small> <a href="https://orcid.org/0000-0001-9450-5023" target="orcid.widget" aria-label="ORCID" class="external-link"><span class="fab fa-orcid orcid" aria-hidden="true"></span></a>  </li>
</ul>
</div>

<div class="dev-status">
<h2 data-toc-skip>Dev status</h2>
<ul class="list-unstyled">
<li><a href="https://lifecycle.r-lib.org/articles/stages.html" class="external-link"><img src="https://img.shields.io/badge/lifecycle-stable-brightgreen.svg" alt="Lifecycle: stable"></a></li>
<li><a href="https://github.com/dcgerard/ldsep/actions/workflows/R-CMD-check.yaml" class="external-link"><img src="https://github.com/dcgerard/ldsep/actions/workflows/R-CMD-check.yaml/badge.svg" alt="R-CMD-check"></a></li>
<li><a href="https://app.codecov.io/gh/dcgerard/ldsep" class="external-link"><img src="https://codecov.io/gh/dcgerard/ldsep/graph/badge.svg" alt="Codecov test coverage"></a></li>
<li><a href="https://CRAN.R-project.org/package=ldsep" class="external-link"><img src="https://www.r-pkg.org/badges/version/ldsep" alt="CRAN status"></a></li>
<li><a href="https://cran.r-project.org/package=ldsep" class="external-link"><img src="http://cranlogs.r-pkg.org/badges/grand-total/ldsep"></a></li>
<li><a href="https://www.gnu.org/licenses/gpl-3.0" class="external-link"><img src="https://img.shields.io/badge/License-GPL%20v3-blue.svg" alt="License: GPL v3"></a></li>
</ul>
</div>

  </aside>
</div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by David Gerard.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.3.</p>
</div>

    </footer>
</div>





  </body>
</html>
