<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Fast LD Estimation • ldsep</title>
<!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png">
<link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png">
<link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png">
<link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png">
<link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png">
<link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png">
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Fast LD Estimation">
<meta property="og:description" content="ldsep">
<meta property="og:image" content="https://dcgerard.github.io/ldsep/logo.png">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">ldsep</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">2.0.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../index.html">
    <span class="fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/fast.html">Fast LD Estimation</a>
    </li>
    <li>
      <a href="../articles/vcf.html">LD Estimation Pipeline from VCF</a>
    </li>
  </ul>
</li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/dcgerard/ldsep/">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><script src="fast_files/header-attrs-2.6/header-attrs.js"></script><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>Fast LD Estimation</h1>
                        <h4 class="author">David Gerard</h4>
            
      
      <small class="dont-index">Source: <a href="https://github.com/dcgerard/ldsep/blob/master/vignettes/fast.Rmd"><code>vignettes/fast.Rmd</code></a></small>
      <div class="hidden name"><code>fast.Rmd</code></div>

    </div>

    
    
<div id="abstract" class="section level1">
<h1 class="hasAnchor">
<a href="#abstract" class="anchor"></a>Abstract</h1>
<p>We demonstrate our new moment-based linkage disequilibrium (LD) estimators, as implemented in the <code><a href="../reference/ldfast.html">ldfast()</a></code> function. The following code chunks are only evaluated if <code>updog</code> is installed with a version of at least 2.0.2.</p>
</div>
<div id="analysis" class="section level1">
<h1 class="hasAnchor">
<a href="#analysis" class="anchor"></a>Analysis</h1>
<p>We will load the packages and data that we will need</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="http://github.com/dcgerard/updog/">updog</a></span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va">ldsep</span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/taiyun/corrplot">corrplot</a></span><span class="op">)</span>
<span class="co">#&gt; corrplot 0.84 loaded</span>
<span class="fu"><a href="https://rdrr.io/r/utils/data.html">data</a></span><span class="op">(</span><span class="st">"uit"</span>, package <span class="op">=</span> <span class="st">"ldsep"</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/base/class.html">class</a></span><span class="op">(</span><span class="va">uit</span><span class="op">)</span>
<span class="co">#&gt; [1] "multidog"</span></code></pre></div>
<p>We need the posterior probabilities for each genotype for each individual at each SNP. In <code>updog</code>, this corresponds to the <code>Pr_n</code> variables in its output, where <code>n</code> varies from 0 to the ploidy. We can use <code><a href="https://rdrr.io/pkg/updog/man/format_multidog.html">updog::format_multidog()</a></code> to extract these values into an array with the proper dimensions.</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">ploidy</span> <span class="op">&lt;-</span> <span class="fl">4</span>
<span class="va">gp</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/updog/man/format_multidog.html">format_multidog</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">uit</span>, varname <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste0</a></span><span class="op">(</span><span class="st">"Pr_"</span>, <span class="fl">0</span><span class="op">:</span><span class="va">ploidy</span><span class="op">)</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/base/class.html">class</a></span><span class="op">(</span><span class="va">gp</span><span class="op">)</span>
<span class="co">#&gt; [1] "array"</span>
<span class="fu"><a href="https://rdrr.io/r/base/dim.html">dim</a></span><span class="op">(</span><span class="va">gp</span><span class="op">)</span>
<span class="co">#&gt; [1] 10 84  5</span></code></pre></div>
<p>The first dimension indexes the SNPs, the second dimension indexes the individuals, and the third dimension indexes the possible dosages. So <code>gp[i, j, k]</code> contains the posterior probability that individual <code>j</code> has dosage <code>k-1</code> at SNP <code>i</code>. You can use other genotyping programs to obtain these posterior probabilities, but they have to be formatted as an array with the dimensions in the correct order (SNPs by individuals by dosages).</p>
<p>If you only have genotype log-likelihoods, we can normalize these to posterior probabilities (assuming a uniform prior) with <code><a href="../reference/gl_to_gp.html">ldsep::gl_to_gp()</a></code>.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">gl</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/updog/man/format_multidog.html">format_multidog</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">uit</span>, varname <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste0</a></span><span class="op">(</span><span class="st">"logL_"</span>, <span class="fl">0</span><span class="op">:</span><span class="va">ploidy</span><span class="op">)</span><span class="op">)</span>
<span class="va">gp2</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/gl_to_gp.html">gl_to_gp</a></span><span class="op">(</span>gl <span class="op">=</span> <span class="va">gl</span><span class="op">)</span></code></pre></div>
<p>Let’s use our original genotype posterior probabilities (<code>gp</code>) to estimate LD since the prior used there was adaptive to the data (but for large sample sizes this does not really matter).</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">ldout</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/ldfast.html">ldfast</a></span><span class="op">(</span>gp <span class="op">=</span> <span class="va">gp</span>, type <span class="op">=</span> <span class="st">"r2"</span><span class="op">)</span></code></pre></div>
<p><code>ldout</code> contains:</p>
<ul>
<li>
<p><code>ldmat</code>: The estimated squared correlation coefficients, adjusted for genotype uncertainty.</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/pkg/corrplot/man/corrplot.html">corrplot</a></span><span class="op">(</span>corr <span class="op">=</span> <span class="va">ldout</span><span class="op">$</span><span class="va">ldmat</span>, 
         method <span class="op">=</span> <span class="st">"color"</span>, 
         type <span class="op">=</span> <span class="st">"upper"</span>,
         diag <span class="op">=</span> <span class="cn">FALSE</span>,
         tl.pos <span class="op">=</span> <span class="st">"n"</span><span class="op">)</span></code></pre></div>
<p><img src="fast_files/figure-html/unnamed-chunk-5-1.png" width="700"></p>
</li>
<li>
<p><code>semat</code>: The estimated standard errors for the squared correlation coefficient estimates.</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/pkg/corrplot/man/corrplot.html">corrplot</a></span><span class="op">(</span>corr <span class="op">=</span> <span class="va">ldout</span><span class="op">$</span><span class="va">semat</span>, 
         method <span class="op">=</span> <span class="st">"color"</span>, 
         type <span class="op">=</span> <span class="st">"upper"</span>,
         diag <span class="op">=</span> <span class="cn">FALSE</span>,
         tl.pos <span class="op">=</span> <span class="st">"n"</span>, 
         is.corr <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></code></pre></div>
<p><img src="fast_files/figure-html/unnamed-chunk-6-1.png" width="700"></p>
</li>
<li>
<p><code>rr</code>: The estimated “reliability ratios”, the multiplicative effects applied to the naive LD estimates using just the posterior mean genotypes.</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu">graphics</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/r/graphics/hist.html">hist</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">ldout</span><span class="op">$</span><span class="va">rr</span>, 
               main <span class="op">=</span> <span class="st">"Reliability Ratios"</span>, 
               xlab <span class="op">=</span> <span class="st">"Reliability Ratios"</span><span class="op">)</span></code></pre></div>
<p><img src="fast_files/figure-html/unnamed-chunk-7-1.png" width="700"></p>
</li>
</ul>
<p>The slowest part of <code><a href="../reference/ldfast.html">ldfast()</a></code> is using hierarchical shrinkage on the log of the reliability ratios (<code>shrinkrr = TRUE</code>), but I wouldn’t disable this as some reliability ratios have high variance and are greatly improved through shrinkage.</p>
<p>Other LD estimators are available: <code>D</code>, <code>Dprime</code>, <code>r</code>, and <code>z</code>. See the help file of <code><a href="../reference/ldfast.html">ldfast()</a></code> for details.</p>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p>Developed by David Gerard.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.6.1.</p>
</div>

      </footer>
</div>

  


  </body>
</html>
