<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Example Pipeline for Estimating LD • ldsep</title>
<!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png">
<link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png">
<link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png">
<link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png">
<link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png">
<link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png">
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Example Pipeline for Estimating LD">
<meta property="og:description" content="ldsep">
<meta property="og:image" content="https://dcgerard.github.io/ldsep/logo.png">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">ldsep</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">2.0.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../index.html">
    <span class="fas fa fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/fast.html">Fast LD Estimation</a>
    </li>
    <li>
      <a href="../articles/ld_pipeline.html">Example Pipeline for Estimating LD</a>
    </li>
  </ul>
</li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/dcgerard/ldsep/">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><script src="ld_pipeline_files/accessible-code-block-0.0.1/empty-anchor.js"></script><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>Example Pipeline for Estimating LD</h1>
                        <h4 class="author">David Gerard</h4>
            
      
      <small class="dont-index">Source: <a href="https://github.com/dcgerard/ldsep/blob/master/vignettes/ld_pipeline.Rmd"><code>vignettes/ld_pipeline.Rmd</code></a></small>
      <div class="hidden name"><code>ld_pipeline.Rmd</code></div>

    </div>

    
    
<div id="abstract" class="section level1">
<h1 class="hasAnchor">
<a href="#abstract" class="anchor"></a>Abstract</h1>
<p>We use the data from Uitdewilligen et. al. (2013) to demonstrate how to obtain LD estimates using the ldsep package. The R chunks are only evaluated if updog (&gt;= 2.0.2) is installed.</p>
</div>
<div id="analysis" class="section level1">
<h1 class="hasAnchor">
<a href="#analysis" class="anchor"></a>Analysis</h1>
<p>We will first load the ldsep package. We will also need to load the updog (Gerard et. al., 2018; Gerard and Ferrão, 2020) and corrplot (Wei and Simko, 2017) R packages for this vignette:</p>
<div class="sourceCode" id="cb1"><pre class="downlit">
<span class="fu"><a href="https://rdrr.io/r/base/library.html">library</a></span>(<span class="kw">ldsep</span>)
<span class="fu"><a href="https://rdrr.io/r/base/library.html">library</a></span>(<span class="kw">updog</span>)
<span class="fu"><a href="https://rdrr.io/r/base/library.html">library</a></span>(<span class="kw"><a href="https://github.com/taiyun/corrplot">corrplot</a></span>)
</pre></div>
<p>The data come from 10 (mostly arbitrary) SNPs from the tetraploid potato dataset of Uitdewilligen et. al. (2013). These SNPs are located close to each other on the same super scaffold. We genotyped these SNPs using the <code><a href="https://rdrr.io/pkg/updog/man/multidog.html">multidog()</a></code> function from the updog R package. The <code>uit</code> object is the resulting output.</p>
<div class="sourceCode" id="cb2"><pre class="downlit">
<span class="kw">ploidy</span> <span class="op">&lt;-</span> <span class="fl">4</span>
<span class="fu"><a href="https://rdrr.io/r/utils/data.html">data</a></span>(<span class="st">"uit"</span>, package = <span class="st">"ldsep"</span>)
<span class="fu"><a href="https://rdrr.io/r/base/class.html">class</a></span>(<span class="kw">uit</span>)
<span class="co">#&gt; [1] "multidog"</span>
</pre></div>
<p>In order to estimate LD, we need to extract the genotype <em>log</em>-likelihoods from the updog fits using <code><a href="https://rdrr.io/pkg/updog/man/format_multidog.html">format_multidog()</a></code> from the updog package.</p>
<div class="sourceCode" id="cb3"><pre class="downlit">
<span class="kw">vvec</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste0</a></span>(<span class="st">"logL_"</span>, <span class="fl">0</span><span class="op">:</span><span class="kw">ploidy</span>)
<span class="kw">vvec</span>
<span class="co">#&gt; [1] "logL_0" "logL_1" "logL_2" "logL_3" "logL_4"</span>
<span class="kw">genoarray</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/updog/man/format_multidog.html">format_multidog</a></span>(x = <span class="kw">uit</span>, varname = <span class="kw">vvec</span>)
<span class="fu"><a href="https://rdrr.io/r/base/dim.html">dim</a></span>(<span class="kw">genoarray</span>)
<span class="co">#&gt; [1] 10 84  5</span>
</pre></div>
<p>The array, <code>genoarray</code>, contains the genotype log-likelihoods with dimensions (SNPs <span class="math inline">\(\times\)</span> individuals <span class="math inline">\(\times\)</span> dosage). So <code>genoarray[i, j, k]</code> contains the genotype log-likelihood of SNP <code>i</code> for individual <code>j</code> at dosage <code>k-1</code>.</p>
<p>It is not necessary to use the genotype log-likelihoods from updog. There are many other genotyping methods that produce genotype log-likelihoods and these could be used instead, as long as you have formatted the genotype log-likelihoods in terms of an array as above.</p>
<p>Let’s estimate all pairwise LD using these genotype log-likelihoods:</p>
<div class="sourceCode" id="cb4"><pre class="downlit">
<span class="kw">ldout</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/mldest.html">mldest</a></span>(geno = <span class="kw">genoarray</span>, K = <span class="kw">ploidy</span>, type = <span class="st">"gam"</span>)
</pre></div>
<p>We can visualize LD with <code><a href="../reference/plot.lddf.html">plot.lddf()</a></code>.</p>
<div class="sourceCode" id="cb5"><pre class="downlit">
<span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html">plot</a></span>(x = <span class="kw">ldout</span>, element = <span class="st">"D"</span>)
</pre></div>
<p><img src="ld_pipeline_files/figure-html/unnamed-chunk-5-1.png" width="700"></p>
<div class="sourceCode" id="cb6"><pre class="downlit">
<span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html">plot</a></span>(x = <span class="kw">ldout</span>, element = <span class="st">"Dprime"</span>)
</pre></div>
<p><img src="ld_pipeline_files/figure-html/unnamed-chunk-5-2.png" width="700"></p>
<div class="sourceCode" id="cb7"><pre class="downlit">
<span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html">plot</a></span>(x = <span class="kw">ldout</span>, element = <span class="st">"r"</span>)
</pre></div>
<p><img src="ld_pipeline_files/figure-html/unnamed-chunk-5-3.png" width="700"></p>
<p>We can also obtain these LD estimates in matrix form by using <code><a href="../reference/format_lddf.html">format_lddf()</a></code>.</p>
<div class="sourceCode" id="cb8"><pre class="downlit">
<span class="kw">ldmat</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/format_lddf.html">format_lddf</a></span>(obj = <span class="kw">ldout</span>, element = <span class="st">"D"</span>)
<span class="kw">ldmat</span>[<span class="fl">1</span><span class="op">:</span><span class="fl">4</span>, <span class="fl">1</span><span class="op">:</span><span class="fl">4</span>]
<span class="co">#&gt;               PotVar0029907 PotVar0029908 PotVar0029910 PotVar0029911</span>
<span class="co">#&gt; PotVar0029907            NA     0.2449746    0.09149879     0.1996393</span>
<span class="co">#&gt; PotVar0029908            NA            NA    0.09149879     0.1996393</span>
<span class="co">#&gt; PotVar0029910            NA            NA            NA     0.1260232</span>
<span class="co">#&gt; PotVar0029911            NA            NA            NA            NA</span>
</pre></div>
<p>Though not generally recommended, we get very similar performance in this case when we use posterior mean genotypes to obtain LD estimates.</p>
<div class="sourceCode" id="cb9"><pre class="downlit">
<span class="kw">pmmat</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/updog/man/format_multidog.html">format_multidog</a></span>(x = <span class="kw">uit</span>, varname = <span class="st">"postmean"</span>)
<span class="kw">ld_pm</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/mldest.html">mldest</a></span>(geno = <span class="kw">pmmat</span>, K = <span class="kw">ploidy</span>, type = <span class="st">"comp"</span>)
<span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html">plot</a></span>(x = <span class="kw">ld_pm</span>, element = <span class="st">"r"</span>)
</pre></div>
<p><img src="ld_pipeline_files/figure-html/unnamed-chunk-7-1.png" width="700"></p>
<div class="sourceCode" id="cb10"><pre class="downlit">
<span class="kw">r_hap</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/format_lddf.html">format_lddf</a></span>(obj = <span class="kw">ldout</span>, element = <span class="st">"r"</span>)
<span class="kw">r_pm</span>  <span class="op">&lt;-</span> <span class="fu"><a href="../reference/format_lddf.html">format_lddf</a></span>(obj = <span class="kw">ld_pm</span>, element = <span class="st">"r"</span>)
<span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html">plot</a></span>(x = <span class="kw">r_hap</span>, 
     y = <span class="kw">r_pm</span>, 
     xlab = <span class="fu"><a href="https://rdrr.io/r/base/expression.html">expression</a></span>(<span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste</a></span>(<span class="fu"><a href="https://rdrr.io/r/grDevices/plotmath.html">hat</a></span>(<span class="kw">r</span>), <span class="st">" Using Genotype Likelihoods"</span>)), 
     ylab = <span class="fu"><a href="https://rdrr.io/r/base/expression.html">expression</a></span>(<span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste</a></span>(<span class="fu"><a href="https://rdrr.io/r/grDevices/plotmath.html">hat</a></span>(<span class="kw">r</span>), <span class="st">" Using Posterior Mean"</span>)))
<span class="fu"><a href="https://rdrr.io/r/graphics/abline.html">abline</a></span>(<span class="fl">0</span>, <span class="fl">1</span>, lty = <span class="fl">2</span>, col = <span class="fl">2</span>)
</pre></div>
<p><img src="ld_pipeline_files/figure-html/unnamed-chunk-8-1.png" width="700"></p>
<p>We can obtain shrunken LD estimates with <code><a href="../reference/ldshrink.html">ldshrink()</a></code>. This shrinkage uses adaptive shrinkage (Stephens, 2016) as proposed in Dey and Stephens (2018).</p>
<div class="sourceCode" id="cb11"><pre class="downlit">
<span class="kw">shrunkcor</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/ldshrink.html">ldshrink</a></span>(obj = <span class="kw">ldout</span>)
<span class="fu"><a href="https://rdrr.io/pkg/corrplot/man/corrplot.html">corrplot</a></span>(corr    = <span class="kw">shrunkcor</span>, 
         type    = <span class="st">"upper"</span>, 
         method  = <span class="st">"color"</span>, 
         diag    = <span class="fl">FALSE</span>, 
         is.corr = <span class="fl">TRUE</span>, 
         tl.pos  = <span class="st">"n"</span>)
</pre></div>
<p><img src="ld_pipeline_files/figure-html/unnamed-chunk-9-1.png" width="700"></p>
<p>To estimate pairwise LD along a sliding window, use <code><a href="../reference/sldest.html">sldest()</a></code>:</p>
<div class="sourceCode" id="cb12"><pre class="downlit">
<span class="kw">slout</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/sldest.html">sldest</a></span>(geno = <span class="kw">genoarray</span>, K = <span class="kw">ploidy</span>, win = <span class="fl">3</span>)
<span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html">plot</a></span>(<span class="kw">slout</span>, element = <span class="st">"r"</span>)
</pre></div>
<p><img src="ld_pipeline_files/figure-html/unnamed-chunk-10-1.png" width="700"></p>
</div>
<div id="references" class="section level1">
<h1 class="hasAnchor">
<a href="#references" class="anchor"></a>References</h1>
<ul>
<li>Dey, Kushal K., and Matthew Stephens. “CorShrink: Empirical Bayes shrinkage estimation of correlations, with applications.” <em>bioRxiv</em> (2018): 368316. <a href="https://doi.org/10.1101/368316">DOI:10.1101/368316</a>
</li>
<li>Gerard, David, Luís Felipe Ventorim Ferrão, Antonio Augusto Franco Garcia, and Matthew Stephens. “Genotyping polyploids from messy sequencing data.” <em>Genetics</em> 210, no. 3 (2018): 789-807. <a href="https://doi.org/10.1534/genetics.118.301468">DOI:10.1534/genetics.118.301468</a>
</li>
<li>Gerard, David, and Luís Felipe Ventorim Ferrão. “Priors for genotyping polyploids.” Bioinformatics 36, no. 6 (2020): 1795-1800. <a href="https://doi.org/10.1093/bioinformatics/btz852">DOI:10.1093/bioinformatics/btz852</a>
</li>
<li>Stephens, Matthew. “False discovery rates: a new deal.” Biostatistics 18, no. 2 (2017): 275-294. <a href="https://doi.org/10.1093/biostatistics/kxw041">DOI;10.1093/biostatistics/kxw041</a>
</li>
<li>Uitdewilligen, Jan GAML, Anne-Marie A. Wolters, B. Bjorn, Theo JA Borm, Richard GF Visser, and Herman J. Van Eck. “A next-generation sequencing method for genotyping-by-sequencing of highly heterozygous autotetraploid potato.” <em>PloS one</em> 8, no. 5 (2013): e62355. <a href="https://doi.org/10.1371/journal.pone.0062355">DOI:10.1371/journal.pone.0062355</a>
</li>
<li>Taiyun Wei and Viliam Simko (2017). R package “corrplot”: Visualization of a Correlation Matrix (Version 0.84). Available from <a href="https://github.com/taiyun/corrplot" class="uri">https://github.com/taiyun/corrplot</a>
</li>
</ul>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p>Developed by David Gerard.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.6.1.</p>
</div>

      </footer>
</div>

  


  </body>
</html>
