<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1.0"><title>Estimate prior genotype probabilities by maximum likelihood — est_prior_mat • ldsep</title><!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png"><link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png"><link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png"><link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png"><link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png"><link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png"><!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous"><script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css"><script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet"><script src="../pkgdown.js"></script><meta property="og:title" content="Estimate prior genotype probabilities by maximum likelihood — est_prior_mat"><meta property="og:description" content="We maximize the marginal likelihood to obtain estimates of the prior
probabilities of each genotype at each locus. The general categorical
class uses the method of Li (2011). The proportional normal class uses
the method of Gerard and Ferrão (2020)."><meta property="og:image" content="https://dcgerard.github.io/ldsep/logo.png"><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body data-spy="scroll" data-target="#toc">
    

    <div class="container template-reference-topic">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">ldsep</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">2.2.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav"><li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu"><li>
      <a href="../articles/fast.html">Fast LD Estimation</a>
    </li>
    <li>
      <a href="../articles/vcf.html">LD Estimation Pipeline from VCF</a>
    </li>
  </ul></li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul><ul class="nav navbar-nav navbar-right"><li>
  <a href="https://github.com/dcgerard/ldsep/" class="external-link">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul></div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Estimate prior genotype probabilities by maximum likelihood</h1>
    <small class="dont-index">Source: <a href="https://github.com/dcgerard/ldsep/blob/HEAD/R/fast.R" class="external-link"><code>R/fast.R</code></a></small>
    <div class="hidden name"><code>est_prior_mat.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>We maximize the marginal likelihood to obtain estimates of the prior
probabilities of each genotype at each locus. The general categorical
class uses the method of Li (2011). The proportional normal class uses
the method of Gerard and Ferrão (2020).</p>
    </div>

    <div id="ref-usage">
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">est_prior_mat</span><span class="op">(</span><span class="va">gl</span>, method <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"pnorm"</span>, <span class="st">"general"</span><span class="op">)</span>, log <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
    </div>

    <div id="arguments">
    <h2>Arguments</h2>
    <dl><dt>gl</dt>
<dd><p>A three dimensional array of genotype <em>log</em>-likelihoods.
Element <code>gl[i, j, k]</code> is the genotype log-likelihood of dosage
<code>k</code> for individual <code>j</code> at SNP <code>i</code>.</p></dd>


<dt>method</dt>
<dd><p>Should we assume the class of general categorical prior
distributions (<code>"general"</code>) or the class of proportional
normal distributions (<code>"pnorm"</code>)?</p></dd>


<dt>log</dt>
<dd><p>A logical. Should we log the results or not?</p></dd>

</dl></div>
    <div id="value">
    <h2>Value</h2>
    

<p>A matrix of prior genotype probabilities. Element <code>(i, k)</code></p>


<p>is the prior probability of genotype <code>k</code> at locus <code>i</code>.</p>
    </div>
    <div id="references">
    <h2>References</h2>
    
<ul><li><p>Gerard, D., &amp; Ferrão, L. F. V. (2020). Priors for genotyping polyploids. Bioinformatics, 36(6), 1795-1800. <a href="https://doi.org/10.1093/bioinformatics/btz852" class="external-link">doi:10.1093/bioinformatics/btz852</a></p></li>
<li><p>Li, H. (2011). A statistical framework for SNP calling, mutation discovery, association mapping and population genetical parameter estimation from sequencing data. <em>Bioinformatics</em>, 27(21), 2987-2993. <a href="https://doi.org/10.1093/bioinformatics/btr509" class="external-link">doi:10.1093/bioinformatics/btr509</a></p></li>
</ul></div>
    <div id="see-also">
    <h2>See also</h2>
    <div class="dont-index">
<dl><dt><code><a href="gl_to_gp.html">gl_to_gp</a>()</code></dt>
<dd><p>For converting genotype likelihoods to
genotype posteriors using prior matrices.</p></dd>


</dl></div>
    </div>
    <div id="author">
    <h2>Author</h2>
    <p>David Gerard and D. Thomas Scartz</p>
    </div>

    <div id="ref-examples">
    <h2>Examples</h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="st">"glike"</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="fu">est_prior_mat</span><span class="op">(</span><span class="va">glike</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span>             [,1]       [,2]       [,3]       [,4]        [,5]</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  [1,]  -3.213015  -1.616097 -0.9568160 -1.2351711  -2.4511626</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  [2,]  -3.213015  -1.616097 -0.9568160 -1.2351711  -2.4511626</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  [3,]  -3.813907  -2.369151 -1.4508240 -1.0589247  -1.1934534</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  [4,]  -2.611388  -1.189717 -0.8852265 -1.6979167  -3.6277874</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  [5,]  -4.547391  -2.863992 -1.6980130 -1.0494556  -0.9183193</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  [6,]  -4.547391  -2.863992 -1.6980130 -1.0494556  -0.9183193</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  [7,]  -3.747561  -2.338272 -1.4446214 -1.0666088  -1.2042344</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  [8,]  -3.747561  -2.338272 -1.4446214 -1.0666088  -1.2042344</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  [9,] -43.257699 -14.321332 -0.3273840 -1.2758538 -17.1667416</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [10,]  -5.251005  -3.082684 -1.6499255 -0.9527288  -0.9910942</span>
<span class="r-in"><span></span></span>
</code></pre></div>
    </div>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top"><h2 data-toc-skip>Contents</h2>
    </nav></div>
</div>


      <footer><div class="copyright">
  <p></p><p>Developed by David Gerard, Hanwei Hu, D. Thomas Scartz, NSF DBI 2132247.</p>
</div>

<div class="pkgdown">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.5.</p>
</div>

      </footer></div>

  


  

  </body></html>

