<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1.0"><title>A function to simulate genotype log-likelihoods under a given correlation under HWE for two loci — glsim_pairwise • ldsep</title><!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png"><link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png"><link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png"><link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png"><link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png"><link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png"><!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous"><script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css"><script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet"><script src="../pkgdown.js"></script><meta property="og:title" content="A function to simulate genotype log-likelihoods under a given correlation under HWE for two loci — glsim_pairwise"><meta property="og:description" content='A user inputs the correlation (rho), the number of individuals
(nind), the ploidy (ploidy), the allele frequencies at two
loci (pA and pB), and the "read-depth" (larger corresponds
to being more sure of the genotypes) (rdepth). Then this
function will  simulate some genotype log-likelihoods.'><meta property="og:image" content="https://dcgerard.github.io/ldsep/logo.png"><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body data-spy="scroll" data-target="#toc">
    

    <div class="container template-reference-topic">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">ldsep</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">2.2.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav"><li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu"><li>
      <a href="../articles/fast.html">Fast LD Estimation</a>
    </li>
    <li>
      <a href="../articles/vcf.html">LD Estimation Pipeline from VCF</a>
    </li>
  </ul></li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul><ul class="nav navbar-nav navbar-right"><li>
  <a href="https://github.com/dcgerard/ldsep/" class="external-link">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul></div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>A function to simulate genotype log-likelihoods under a given correlation under HWE for two loci</h1>
    <small class="dont-index">Source: <a href="https://github.com/dcgerard/ldsep/blob/HEAD/R/glsim.R" class="external-link"><code>R/glsim.R</code></a></small>
    <div class="hidden name"><code>glsim_pairwise.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>A user inputs the correlation (<code>rho</code>), the number of individuals
(<code>nind</code>), the ploidy (<code>ploidy</code>), the allele frequencies at two
loci (<code>pA</code> and <code>pB</code>), and the "read-depth" (larger corresponds
to being more sure of the genotypes) (<code>rdepth</code>). Then this
function will  simulate some genotype log-likelihoods.</p>
    </div>

    <div id="ref-usage">
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">glsim_pairwise</span><span class="op">(</span></span>
<span>  <span class="va">rho</span>,</span>
<span>  <span class="va">nind</span>,</span>
<span>  <span class="va">ploidy</span>,</span>
<span>  <span class="va">pA</span>,</span>
<span>  <span class="va">pB</span>,</span>
<span>  rdepth <span class="op">=</span> <span class="fl">10</span>,</span>
<span>  seq <span class="op">=</span> <span class="fl">0.001</span>,</span>
<span>  bias <span class="op">=</span> <span class="fl">1</span>,</span>
<span>  od <span class="op">=</span> <span class="fl">0.001</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div id="arguments">
    <h2>Arguments</h2>
    <dl><dt>rho</dt>
<dd><p>The true correlation.</p></dd>


<dt>nind</dt>
<dd><p>The number of individuals to simulate.</p></dd>


<dt>ploidy</dt>
<dd><p>The ploidy of the species.</p></dd>


<dt>pA</dt>
<dd><p>The allele frequency of the first locus.</p></dd>


<dt>pB</dt>
<dd><p>The allele frequency of the second locus.</p></dd>


<dt>rdepth</dt>
<dd><p>The read depth to simulate (larger means more certain).</p></dd>


<dt>seq</dt>
<dd><p>The sequencing error rate.</p></dd>


<dt>bias</dt>
<dd><p>The allele bias.</p></dd>


<dt>od</dt>
<dd><p>The overdispersion parameter.</p></dd>

</dl></div>
    <div id="value">
    <h2>Value</h2>
    

<p>A list with some or all of the following elements</p><dl><dt>lA</dt>
<dd><p>The genotype log-likelihoods at the first locus. <code>lA[i, k]</code>
is the genotype log-likelihood for individual <code>i</code> at genotype <code>k - 1</code>.</p></dd>

<dt>lB</dt>
<dd><p>The genotype log-likelihoods at the second locus. <code>lB[i, K]</code>
is the genotype log-likelihood for individual <code>i</code> at genotype <code>k - 1</code>.</p></dd>

<dt>rho</dt>
<dd><p>The true Pearson correlation that the user provided.</p></dd>

<dt>delta</dt>
<dd><p>The corresponding LD coefficient.</p></dd>

<dt>gA</dt>
<dd><p>The vector of true genotypes for locus A</p></dd>

<dt>gB</dt>
<dd><p>The vector of true genotypes for locus B</p></dd>

<dt>pA</dt>
<dd><p>The matrix of posterior genotype probabilities for locus A</p></dd>

<dt>pB</dt>
<dd><p>The matrix of posterior genotype probabilities for locus B</p></dd>


</dl></div>
    <div id="author">
    <h2>Author</h2>
    <p>David Gerard</p>
    </div>

    <div id="ref-examples">
    <h2>Examples</h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span><span class="va">glout</span> <span class="op">&lt;-</span> <span class="fu">glsim_pairwise</span><span class="op">(</span>rho <span class="op">=</span> <span class="fl">0.9</span>, nind <span class="op">=</span> <span class="fl">10000</span>, ploidy <span class="op">=</span> <span class="fl">6</span>, pA <span class="op">=</span> <span class="fl">0.5</span>, pB <span class="op">=</span> <span class="fl">0.5</span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/stats/cor.html" class="external-link">cor</a></span><span class="op">(</span><span class="va">glout</span><span class="op">$</span><span class="va">gA</span>, <span class="va">glout</span><span class="op">$</span><span class="va">gB</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] 0.9005799</span>
<span class="r-in"><span></span></span>
</code></pre></div>
    </div>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top"><h2 data-toc-skip>Contents</h2>
    </nav></div>
</div>


      <footer><div class="copyright">
  <p></p><p>Developed by David Gerard, Hanwei Hu, D. Thomas Scartz, NSF DBI 2132247.</p>
</div>

<div class="pkgdown">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.5.</p>
</div>

      </footer></div>

  


  

  </body></html>

