<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1.0"><title>Estimate haplotypic pair-wise LD using either genotypes or genotype
likelihoods. — ldest_hap • ldsep</title><!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png"><link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png"><link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png"><link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png"><link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png"><link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png"><!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous"><script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css"><script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet"><script src="../pkgdown.js"></script><meta property="og:title" content="Estimate haplotypic pair-wise LD using either genotypes or genotype
likelihoods. — ldest_hap"><meta property="og:description" content="Given genotype (allele dosage) or genotype likelihood data
for each individual at a pair of loci, this function will
calculate the maximum likelihood estimates
and their corresponding asymptotic standard errors of some
measures of linkage disequilibrium (LD): D, D', the Pearson correlation,
the squared Pearson correlation, and the Fisher-z transformation of the
Pearson correlation. This function can be used for both
diploids and polyploids."><meta property="og:image" content="https://dcgerard.github.io/ldsep/logo.png"><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body data-spy="scroll" data-target="#toc">
    

    <div class="container template-reference-topic">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">ldsep</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">2.1.4</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav"><li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu"><li>
      <a href="../articles/fast.html">Fast LD Estimation</a>
    </li>
    <li>
      <a href="../articles/vcf.html">LD Estimation Pipeline from VCF</a>
    </li>
  </ul></li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul><ul class="nav navbar-nav navbar-right"><li>
  <a href="https://github.com/dcgerard/ldsep/" class="external-link">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul></div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Estimate haplotypic pair-wise LD using either genotypes or genotype
likelihoods.</h1>
    <small class="dont-index">Source: <a href="https://github.com/dcgerard/ldsep/blob/HEAD/R/ldgeno.R" class="external-link"><code>R/ldgeno.R</code></a></small>
    <div class="hidden name"><code>ldest_hap.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>Given genotype (allele dosage) or genotype likelihood data
for each individual at a pair of loci, this function will
calculate the maximum likelihood estimates
and their corresponding asymptotic standard errors of some
measures of linkage disequilibrium (LD): D, D', the Pearson correlation,
the squared Pearson correlation, and the Fisher-z transformation of the
Pearson correlation. This function can be used for both
diploids and polyploids.</p>
    </div>

    <div id="ref-usage">
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="fu">ldest_hap</span><span class="op">(</span>
  <span class="va">ga</span>,
  <span class="va">gb</span>,
  <span class="va">K</span>,
  reltol <span class="op">=</span> <span class="fl">10</span><span class="op">^</span><span class="op">-</span><span class="fl">8</span>,
  nboot <span class="op">=</span> <span class="fl">100</span>,
  useboot <span class="op">=</span> <span class="cn">FALSE</span>,
  pen <span class="op">=</span> <span class="fl">2</span>,
  grid_init <span class="op">=</span> <span class="cn">FALSE</span>,
  se <span class="op">=</span> <span class="cn">TRUE</span>
<span class="op">)</span></code></pre></div>
    </div>

    <div id="arguments">
    <h2>Arguments</h2>
    <dl><dt>ga</dt>
<dd><p>One of two possible inputs:</p><ol><li><p>A vector of counts, containing the genotypes for each
individual at the first locus. When <code>type = "comp"</code>,
the vector of genotypes may be continuous (e.g. the
posterior mean genotype).</p></li>
<li><p>A matrix of genotype log-likelihoods at the first locus.
The rows index the individuals and the columns index
the genotypes. That is <code>ga[i, j]</code> is the genotype
likelihood of individual <code>i</code> for genotype <code>j-1</code>.</p></li>
</ol></dd>
<dt>gb</dt>
<dd><p>One of two possible inputs:</p><ol><li><p>A vector of counts, containing the genotypes for each
individual at the second locus. When <code>type = "comp"</code>,
the vector of genotypes may be continuous (e.g. the
posterior mean genotype).</p></li>
<li><p>A matrix of genotype log-likelihoods at the second locus.
The rows index the individuals and the columns index
the genotypes. That is <code>gb[i, j]</code> is the genotype
likelihood of individual <code>i</code> for genotype <code>j-1</code>.</p></li>
</ol></dd>
<dt>K</dt>
<dd><p>The ploidy of the species. Assumed to be the same for all
individuals.</p></dd>
<dt>reltol</dt>
<dd><p>The relative tolerance for the stopping criterion.</p></dd>
<dt>nboot</dt>
<dd><p>Sometimes, the MLE standard errors don't exist. So we use
the bootstrap as a backup. <code>nboot</code> specifies the number
of bootstrap iterations.</p></dd>
<dt>useboot</dt>
<dd><p>A logical. Optionally, you may always use the bootstrap
to estimate the standard errors (<code>TRUE</code>). These will be more
accurate but also much slower, so this defaults to <code>FALSE</code>. Only
applicable if using genotype likelihoods.</p></dd>
<dt>pen</dt>
<dd><p>The penalty to be applied to the likelihood. You can think about
this as the prior sample size. Should be greater than 1. Does not
apply if <code>model = "norm"</code>, <code>type = "comp"</code>, and using
genotype likelihoods. Also does not apply when <code>type = "comp"</code>
and using genotypes.</p></dd>
<dt>grid_init</dt>
<dd><p>A logical. Should we initialize the gradient ascent
at a grid of initial values (<code>TRUE</code>) or just initialize
at one value corresponding to the simplex point
<code>rep(0.25, 4)</code> (<code>FALSE</code>)?</p></dd>
<dt>se</dt>
<dd><p>A logical. Should we calculate standard errors (<code>TRUE</code>) or
not (<code>FALSE</code>). Calculating standard errors can be really slow
when <code>type = "comp"</code>, <code>model = "flex"</code>, and when using
genotype likelihoods. Otherwise, standard error calculations
should be pretty fast.</p></dd>
</dl></div>
    <div id="value">
    <h2>Value</h2>
    <p>A vector with some or all of the following elements:</p><dl><dt><code>D</code></dt>
<dd><p>The estimate of the LD coefficient.</p></dd>

<dt><code>D_se</code></dt>
<dd><p>The standard error of the estimate of
the LD coefficient.</p></dd>

<dt><code>r2</code></dt>
<dd><p>The estimate of the squared Pearson correlation.</p></dd>

<dt><code>r2_se</code></dt>
<dd><p>The standard error of the estimate of the
squared Pearson correlation.</p></dd>

<dt><code>r</code></dt>
<dd><p>The estimate of the Pearson correlation.</p></dd>

<dt><code>r_se</code></dt>
<dd><p>The standard error of the estimate of the
Pearson correlation.</p></dd>

<dt><code>Dprime</code></dt>
<dd><p>The estimate of the standardized LD
coefficient. When <code>type</code> = "comp", this corresponds
to the standardization where we fix allele frequencies.</p></dd>

<dt><code>Dprime_se</code></dt>
<dd><p>The standard error of <code>Dprime</code>.</p></dd>

<dt><code>Dprimeg</code></dt>
<dd><p>The estimate of the standardized LD
coefficient. This corresponds to the standardization where
we fix genotype frequencies.</p></dd>

<dt><code>Dprimeg_se</code></dt>
<dd><p>The standard error of <code>Dprimeg</code>.</p></dd>

<dt><code>z</code></dt>
<dd><p>The Fisher-z transformation of <code>r</code>.</p></dd>

<dt><code>z_se</code></dt>
<dd><p>The standard error of the Fisher-z
transformation of <code>r</code>.</p></dd>

<dt><code>p_ab</code></dt>
<dd><p>The estimated haplotype frequency of ab.
Only returned if estimating the haplotypic LD.</p></dd>

<dt><code>p_Ab</code></dt>
<dd><p>The estimated haplotype frequency of Ab.
Only returned if estimating the haplotypic LD.</p></dd>

<dt><code>p_aB</code></dt>
<dd><p>The estimated haplotype frequency of aB.
Only returned if estimating the haplotypic LD.</p></dd>

<dt><code>p_AB</code></dt>
<dd><p>The estimated haplotype frequency of AB.
Only returned if estimating the haplotypic LD.</p></dd>

<dt><code>q_ij</code></dt>
<dd><p>The estimated frequency of genotype i at locus 1
and genotype j at locus 2. Only returned if estimating the
composite LD.</p></dd>

<dt><code>n</code></dt>
<dd><p>The number of individuals used to estimate pairwise LD.</p></dd>


</dl></div>
    <div id="details">
    <h2>Details</h2>
    <p>Let A and a be the reference and alternative alleles, respectively, at
locus 1. Let B and b be the reference and alternative alleles,
respectively, at locus 2. Let paa, pAb, paB, and pAB be the
frequencies of haplotypes ab, Ab, aB, and AB, respectively.
Let pA = pAb + pAB and let pB = paB + pAB
The <code>ldest</code> returns estimates of the following measures
of LD.</p><ul><li><p>D: pAB - pA pB</p></li>
<li><p>D': D / Dmax, where Dmax = min(pA pB, (1 - pA) (1 - pB)) if
D &lt; 0 and Dmax = min(pA (1 - pB), pA (1 - pB)) if D &gt; 0</p></li>
<li><p>r-squared: The squared Pearson correlation,
r^2 = D^2 / (pA (1 - pA) pB (1 - pB))</p></li>
<li><p>r: The Pearson correlation,
r = D / sqrt(pA (1 - pA) pB (1 - pB))</p></li>
</ul><p>Estimates are obtained via maximum likelihood under the assumption
of Hardy-Weinberg equilibrium. The likelihood is calculated by
integrating over the possible haplotypes for each pair of genotypes.</p>
<p>The resulting standard errors are based on the square roots of the inverse of the
negative Fisher-information. This is from standard maximum likelihood
theory. The Fisher-information is known to be biased low, so the actual
standard errors are probably a little bigger for small n (n &lt; 20).
In some cases the Fisher-information matrix is singular, and so we
in these cases we return a bootstrap estimate of the standard error.</p>
<p>The standard error estimate of the squared Pearson correlation is not
valid when r^2 = 0.</p>
<p>In cases where either SNP is estimated to be monoallelic
(<code>pA %in% c(0, 1)</code> or <code>pB %in% c(0, 1)</code>), this function
will return LD estimates of <code>NA</code>.</p>
    </div>
    <div id="author">
    <h2>Author</h2>
    <p>David Gerard</p>
    </div>

    <div id="ref-examples">
    <h2>Examples</h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span class="fu"><a href="https://rdrr.io/r/base/Random.html" class="external-link">set.seed</a></span><span class="op">(</span><span class="fl">1</span><span class="op">)</span></span>
<span class="r-in"><span class="va">n</span> <span class="op">&lt;-</span> <span class="fl">100</span> <span class="co"># sample size</span></span>
<span class="r-in"><span class="va">K</span> <span class="op">&lt;-</span> <span class="fl">6</span> <span class="co"># ploidy</span></span>
<span class="r-in"></span>
<span class="r-in"><span class="co">## generate some fake genotypes when LD = 0.</span></span>
<span class="r-in"><span class="va">ga</span> <span class="op">&lt;-</span> <span class="fu">stats</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/r/stats/Binomial.html" class="external-link">rbinom</a></span><span class="op">(</span>n <span class="op">=</span> <span class="va">n</span>, size <span class="op">=</span> <span class="va">K</span>, prob <span class="op">=</span> <span class="fl">0.5</span><span class="op">)</span></span>
<span class="r-in"><span class="va">gb</span> <span class="op">&lt;-</span> <span class="fu">stats</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/r/stats/Binomial.html" class="external-link">rbinom</a></span><span class="op">(</span>n <span class="op">=</span> <span class="va">n</span>, size <span class="op">=</span> <span class="va">K</span>, prob <span class="op">=</span> <span class="fl">0.5</span><span class="op">)</span></span>
<span class="r-in"><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">ga</span><span class="op">)</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] 2 3 3 5 2 5</span>
<span class="r-in"><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">gb</span><span class="op">)</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] 3 3 2 6 3 2</span>
<span class="r-in"></span>
<span class="r-in"><span class="co">## generate some fake genotype likelihoods when LD = 0.</span></span>
<span class="r-in"><span class="va">gamat</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/t.html" class="external-link">t</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/lapply.html" class="external-link">sapply</a></span><span class="op">(</span><span class="va">ga</span>, <span class="fu">stats</span><span class="fu">::</span><span class="va"><a href="https://rdrr.io/r/stats/Normal.html" class="external-link">dnorm</a></span>, x <span class="op">=</span> <span class="fl">0</span><span class="op">:</span><span class="va">K</span>, sd <span class="op">=</span> <span class="fl">1</span>, log <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span><span class="op">)</span></span>
<span class="r-in"><span class="va">gbmat</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/t.html" class="external-link">t</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/lapply.html" class="external-link">sapply</a></span><span class="op">(</span><span class="va">gb</span>, <span class="fu">stats</span><span class="fu">::</span><span class="va"><a href="https://rdrr.io/r/stats/Normal.html" class="external-link">dnorm</a></span>, x <span class="op">=</span> <span class="fl">0</span><span class="op">:</span><span class="va">K</span>, sd <span class="op">=</span> <span class="fl">1</span>, log <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span><span class="op">)</span></span>
<span class="r-in"><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">gamat</span><span class="op">)</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span>            [,1]      [,2]       [,3]       [,4]      [,5]       [,6]      [,7]</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1,]  -2.918939 -1.418939 -0.9189385 -1.4189385 -2.918939 -5.4189385 -8.918939</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [2,]  -5.418939 -2.918939 -1.4189385 -0.9189385 -1.418939 -2.9189385 -5.418939</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [3,]  -5.418939 -2.918939 -1.4189385 -0.9189385 -1.418939 -2.9189385 -5.418939</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [4,] -13.418939 -8.918939 -5.4189385 -2.9189385 -1.418939 -0.9189385 -1.418939</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [5,]  -2.918939 -1.418939 -0.9189385 -1.4189385 -2.918939 -5.4189385 -8.918939</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [6,] -13.418939 -8.918939 -5.4189385 -2.9189385 -1.418939 -0.9189385 -1.418939</span>
<span class="r-in"><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">gbmat</span><span class="op">)</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span>            [,1]       [,2]       [,3]       [,4]      [,5]      [,6]       [,7]</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1,]  -5.418939  -2.918939 -1.4189385 -0.9189385 -1.418939 -2.918939 -5.4189385</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [2,]  -5.418939  -2.918939 -1.4189385 -0.9189385 -1.418939 -2.918939 -5.4189385</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [3,]  -2.918939  -1.418939 -0.9189385 -1.4189385 -2.918939 -5.418939 -8.9189385</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [4,] -18.918939 -13.418939 -8.9189385 -5.4189385 -2.918939 -1.418939 -0.9189385</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [5,]  -5.418939  -2.918939 -1.4189385 -0.9189385 -1.418939 -2.918939 -5.4189385</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [6,]  -2.918939  -1.418939 -0.9189385 -1.4189385 -2.918939 -5.418939 -8.9189385</span>
<span class="r-in"></span>
<span class="r-in"><span class="co">## Haplotypic LD with genotypes</span></span>
<span class="r-in"><span class="va">ldout1</span> <span class="op">&lt;-</span> <span class="fu">ldest_hap</span><span class="op">(</span>ga <span class="op">=</span> <span class="va">ga</span>,</span>
<span class="r-in">                    gb <span class="op">=</span> <span class="va">gb</span>,</span>
<span class="r-in">                    K <span class="op">=</span> <span class="va">K</span><span class="op">)</span></span>
<span class="r-in"><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">ldout1</span><span class="op">)</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span>            D         D_se           r2        r2_se            r         r_se </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 0.0058365062 0.0286113348 0.0005454794 0.0053480254 0.0233554997 0.1144917775 </span>
<span class="r-in"></span>
<span class="r-in"><span class="co">## Haplotypic LD with genotype likelihoods</span></span>
<span class="r-in"><span class="va">ldout2</span> <span class="op">&lt;-</span> <span class="fu">ldest_hap</span><span class="op">(</span>ga <span class="op">=</span> <span class="va">gamat</span>,</span>
<span class="r-in">                    gb <span class="op">=</span> <span class="va">gbmat</span>,</span>
<span class="r-in">                    K <span class="op">=</span> <span class="va">K</span><span class="op">)</span></span>
<span class="r-in"><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">ldout2</span><span class="op">)</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span>           D        D_se          r2       r2_se           r        r_se </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 0.020950505 0.097754998 0.007028751 0.065592168 0.083837645 0.391185656 </span>
<span class="r-in"></span>
</code></pre></div>
    </div>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top"><h2 data-toc-skip>Contents</h2>
    </nav></div>
</div>


      <footer><div class="copyright">
  <p></p><p>Developed by David Gerard.</p>
</div>

<div class="pkgdown">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.2.</p>
</div>

      </footer></div>

  


  

  </body></html>

