<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1.0"><title>Naive approach to estimate LD using posterior moments. — ldfast_naive • ldsep</title><!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png"><link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png"><link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png"><link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png"><link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png"><link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png"><!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous"><script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css"><script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet"><script src="../pkgdown.js"></script><meta property="og:title" content="Naive approach to estimate LD using posterior moments. — ldfast_naive"><meta property="og:description" content="Here, we just use posterior means as a proxy for the genotypes, and calculate
these LD estimates as if the genotypes were known. This is not a good way
to do things if there is any genotype uncertainty (which is almost always
the case in polyploids)."><meta property="og:image" content="https://dcgerard.github.io/ldsep/logo.png"><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body data-spy="scroll" data-target="#toc">
    

    <div class="container template-reference-topic">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">ldsep</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">2.2.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav"><li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu"><li>
      <a href="../articles/fast.html">Fast LD Estimation</a>
    </li>
    <li>
      <a href="../articles/vcf.html">LD Estimation Pipeline from VCF</a>
    </li>
  </ul></li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul><ul class="nav navbar-nav navbar-right"><li>
  <a href="https://github.com/dcgerard/ldsep/" class="external-link">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul></div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Naive approach to estimate LD using posterior moments.</h1>
    <small class="dont-index">Source: <a href="https://github.com/dcgerard/ldsep/blob/HEAD/R/fast.R" class="external-link"><code>R/fast.R</code></a></small>
    <div class="hidden name"><code>ldfast_naive.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>Here, we just use posterior means as a proxy for the genotypes, and calculate
these LD estimates as if the genotypes were known. This is not a good way
to do things if there is any genotype uncertainty (which is almost always
the case in polyploids).</p>
    </div>

    <div id="ref-usage">
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">ldfast_naive</span><span class="op">(</span><span class="va">gp</span>, type <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"r"</span>, <span class="st">"r2"</span>, <span class="st">"z"</span>, <span class="st">"D"</span>, <span class="st">"Dprime"</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
    </div>

    <div id="arguments">
    <h2>Arguments</h2>
    <dl><dt>gp</dt>
<dd><p>A three-way array with dimensions SNPs by individuals by dosage.
That is, <code>gp[i, j, k]</code> is the posterior probability of
dosage <code>k-1</code> for individual <code>j</code> at SNP <code>i</code>.</p></dd>


<dt>type</dt>
<dd><p>What LD measure should we estimate?</p><dl><dt><code>"r"</code></dt>
<dd><p>The Pearson correlation.</p></dd>

<dt><code>"r2"</code></dt>
<dd><p>The squared Pearson correlation.</p></dd>

<dt><code>"z"</code></dt>
<dd><p>The Fisher-z transformed Pearson correlation.</p></dd>

<dt><code>"D"</code></dt>
<dd><p>The LD coefficient.</p></dd>

<dt><code>"Dprime"</code></dt>
<dd><p>The standardized LD coefficient.</p></dd>


</dl><p>Note that these are all <em>composite</em> measures of LD (see
the description in <code><a href="ldest.html">ldest</a>()</code>).</p></dd>

</dl></div>
    <div id="value">
    <h2>Value</h2>
    

<p>A list with the following elements.</p><dl><dt><code>ldmat</code></dt>
<dd><p>The bias-corrected LD matrix.</p></dd>


</dl></div>
    <div id="author">
    <h2>Author</h2>
    <p>David Gerard</p>
    </div>

    <div id="ref-examples">
    <h2>Examples</h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="st">"glike"</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">gp</span> <span class="op">&lt;-</span> <span class="fu"><a href="gl_to_gp.html">gl_to_gp</a></span><span class="op">(</span>gl <span class="op">=</span> <span class="va">glike</span>, prior_mat <span class="op">=</span> <span class="st">"estimate_pnorm"</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="fu">ldfast_naive</span><span class="op">(</span>gp <span class="op">=</span> <span class="va">gp</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> $ldmat</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>             [,1]       [,2]       [,3]         [,4]      [,5]      [,6]</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  [1,] 1.00000000 1.00000000 0.42862407  0.823237426 0.5983991 0.5983991</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  [2,] 1.00000000 1.00000000 0.42862407  0.823237426 0.5983991 0.5983991</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  [3,] 0.42862407 0.42862407 1.00000000  0.569075315 0.8879786 0.8879786</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  [4,] 0.82323743 0.82323743 0.56907532  1.000000000 0.5454147 0.5454147</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  [5,] 0.59839914 0.59839914 0.88797855  0.545414700 1.0000000 1.0000000</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  [6,] 0.59839914 0.59839914 0.88797855  0.545414700 1.0000000 1.0000000</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  [7,] 0.42404788 0.42404788 0.97149161  0.576466698 0.8835920 0.8835920</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  [8,] 0.42404788 0.42404788 0.97149161  0.576466698 0.8835920 0.8835920</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  [9,] 0.09693651 0.09693651 0.08726337 -0.001911192 0.1260134 0.1260134</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [10,] 0.53656274 0.53656274 0.79518829  0.521588524 0.9288950 0.9288950</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>             [,7]       [,8]         [,9]     [,10]</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  [1,] 0.42404788 0.42404788  0.096936509 0.5365627</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  [2,] 0.42404788 0.42404788  0.096936509 0.5365627</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  [3,] 0.97149161 0.97149161  0.087263374 0.7951883</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  [4,] 0.57646670 0.57646670 -0.001911192 0.5215885</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  [5,] 0.88359197 0.88359197  0.126013372 0.9288950</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  [6,] 0.88359197 0.88359197  0.126013372 0.9288950</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  [7,] 1.00000000 1.00000000  0.074136925 0.8212752</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  [8,] 1.00000000 1.00000000  0.074136925 0.8212752</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  [9,] 0.07413693 0.07413693  1.000000000 0.0930604</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [10,] 0.82127524 0.82127524  0.093060400 1.0000000</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-in"><span></span></span>
</code></pre></div>
    </div>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top"><h2 data-toc-skip>Contents</h2>
    </nav></div>
</div>


      <footer><div class="copyright">
  <p></p><p>Developed by David Gerard, Hanwei Hu, D. Thomas Scartz, NSF DBI 2132247.</p>
</div>

<div class="pkgdown">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.5.</p>
</div>

      </footer></div>

  


  

  </body></html>

