<!-- Generated by pkgdown: do not edit by hand -->
<!DOCTYPE html>
<html lang="en">
  <head>
  <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<title>Fast bias-correction for LD Estimation — ldfast • ldsep</title>

<!-- favicons -->
<link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png">
<link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png">
<link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png" />
<link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png" />
<link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png" />
<link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png" />

<!-- jquery -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script>
<!-- Bootstrap -->

<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous" />

<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script>

<!-- bootstrap-toc -->
<link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script>

<!-- Font Awesome icons -->
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous" />
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous" />

<!-- clipboard.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script>

<!-- headroom.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script>

<!-- pkgdown -->
<link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script>




<meta property="og:title" content="Fast bias-correction for LD Estimation — ldfast" />
<meta property="og:description" content="Estimates the reliability ratios from posterior marginal moments and uses
these to correct the biases in linkage disequilibrium estimation
caused by genotype uncertainty. These methods are described in
Gerard (2021)." />
<meta property="og:image" content="https://dcgerard.github.io/ldsep/logo.png" />




<!-- mathjax -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script>

<!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->



  </head>

  <body data-spy="scroll" data-target="#toc">
    <div class="container template-reference-topic">
      <header>
      <div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">ldsep</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">2.0.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="../index.html">
    <span class="fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="../articles/fast.html">Fast LD Estimation</a>
    </li>
    <li>
      <a href="../articles/vcf.html">LD Estimation Pipeline from VCF</a>
    </li>
  </ul>
</li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/dcgerard/ldsep/">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
      
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header>

<div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Fast bias-correction for LD Estimation</h1>
    <small class="dont-index">Source: <a href='https://github.com/dcgerard/ldsep/blob/master/R/fast.R'><code>R/fast.R</code></a></small>
    <div class="hidden name"><code>ldfast.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>Estimates the reliability ratios from posterior marginal moments and uses
these to correct the biases in linkage disequilibrium estimation
caused by genotype uncertainty. These methods are described in
Gerard (2021).</p>
    </div>

    <pre class="usage"><span class='fu'>ldfast</span><span class='op'>(</span>
  <span class='va'>gp</span>,
  type <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span><span class='op'>(</span><span class='st'>"r"</span>, <span class='st'>"r2"</span>, <span class='st'>"z"</span>, <span class='st'>"D"</span>, <span class='st'>"Dprime"</span><span class='op'>)</span>,
  shrinkrr <span class='op'>=</span> <span class='cn'>TRUE</span>,
  se <span class='op'>=</span> <span class='cn'>TRUE</span>,
  thresh <span class='op'>=</span> <span class='cn'>TRUE</span>,
  upper <span class='op'>=</span> <span class='fl'>10</span>,
  mode <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span><span class='op'>(</span><span class='st'>"zero"</span>, <span class='st'>"estimate"</span><span class='op'>)</span>
<span class='op'>)</span></pre>

    <h2 class="hasAnchor" id="arguments"><a class="anchor" href="#arguments"></a>Arguments</h2>
    <table class="ref-arguments">
    <colgroup><col class="name" /><col class="desc" /></colgroup>
    <tr>
      <th>gp</th>
      <td><p>A three-way array with dimensions SNPs by individuals by dosage.
That is, <code>gp[i, j, k]</code> is the posterior probability of
dosage <code>k-1</code> for individual <code>j</code> at SNP <code>i</code>.</p></td>
    </tr>
    <tr>
      <th>type</th>
      <td><p>What LD measure should we estimate?</p><dl>
<dt><code>"r"</code></dt><dd><p>The Pearson correlation.</p></dd>
<dt><code>"r2"</code></dt><dd><p>The squared Pearson correlation.</p></dd>
<dt><code>"z"</code></dt><dd><p>The Fisher-z transformed Pearson correlation.</p></dd>
<dt><code>"D"</code></dt><dd><p>The LD coefficient.</p></dd>
<dt><code>"Dprime"</code></dt><dd><p>The standardized LD coefficient.</p></dd>

</dl><p>Note that these are all <em>composite</em> measures of LD (see
the description in <code><a href='ldest.html'>ldest</a>()</code>).</p></td>
    </tr>
    <tr>
      <th>shrinkrr</th>
      <td><p>A logical. Should we use adaptive shrinkage
(Stephens, 2016) to shrink the reliability ratios (<code>TRUE</code>)
or keep the raw reliability ratios (<code>FALSE</code>). Defaults
to <code>TRUE</code>.</p></td>
    </tr>
    <tr>
      <th>se</th>
      <td><p>Should we also return a matrix of standard errors (<code>TRUE</code>)
or not (<code>FALSE</code>)? It is faster to not return standard errors.
Defaults to <code>TRUE</code>.</p></td>
    </tr>
    <tr>
      <th>thresh</th>
      <td><p>A logical. Should we apply an upper bound on the reliability
ratios (<code>TRUE</code>) or not (<code>FALSE</code>).</p></td>
    </tr>
    <tr>
      <th>upper</th>
      <td><p>The upper bound on the reliability ratios if
<code>thresh = TRUE</code>. The default is a generous 10.</p></td>
    </tr>
    <tr>
      <th>mode</th>
      <td><p>A character. Only applies if <code>shrinkrr = TRUE</code>. When using
hierarchical shrinkage on the log of the reliability ratios, should
we use zero as the mode (<code>mode = "zero"</code>) or estimate it using
the procedure of Robertson and Cryer (1974)
(<code>mode = "estimate"</code>)?</p></td>
    </tr>
    </table>

    <h2 class="hasAnchor" id="value"><a class="anchor" href="#value"></a>Value</h2>

    <p>A list with some or all of the following elements:</p><dl>
<dt><code>ldmat</code></dt><dd><p>The bias-corrected LD matrix.</p></dd>
<dt><code>rr</code></dt><dd><p>The estimated reliability ratio for each SNP. This
is the multiplicative factor applied to the naive LD estimate
for each SNP.</p></dd>
<dt><code>semat</code></dt><dd><p>A matrix of standard errors of the corresponding
estimators of LD.</p></dd>

</dl>

    <h2 class="hasAnchor" id="details"><a class="anchor" href="#details"></a>Details</h2>

    


<p>Returns consistent and bias-corrected estimates of linkage disequilibrium.
The usual measures of LD are implemented: D, D', r, r2, and z
(Fisher-z of r). These are all <em>composite</em> measures of LD, not
haplotypic measures of LD (see the description in <code><a href='ldest.html'>ldest</a>()</code>).
They are always appropriate measures of association
between loci, but only correspond to haplotypic measures of LD when
Hardy-Weinberg equilibrium is fulfilled in autopolyploids.</p>
<p>Calculating standard errors and performing hierarchical shrinkage of the
reliability ratios are both rather slow operations compared to just
raw method-of-moments based estimation for LD. If you don't need
standard errors, you can double your speed by setting
<code>se = FALSE</code>. It is not recommended that you disable the
hierarchical shrinkage.</p>
<p>Due to sampling variability, the estimates sometime lie outside of the
theoretical boundaries of the parameters being estimated. In such cases,
we truncate the estimates at the boundary and return <code>NA</code> for the
standard errors.</p>
    <h2 class="hasAnchor" id="mathematical-formulation"><a class="anchor" href="#mathematical-formulation"></a>Mathematical formulation</h2>

    

<p>Let</p><ul>
<li><p>\(r\) be the sample correlation of posterior mean genotypes
between loci 1 and 2,</p></li>
<li><p>\(a1\) be the sample variance of posterior means at locus 1,</p></li>
<li><p>\(a2\) be the sample variance of posterior means at locus 2,</p></li>
<li><p>\(b1\) be the sample mean of posterior variances at locus 1, and</p></li>
<li><p>\(b2\) be the sample mean of posterior variances at locus 2.</p></li>
</ul><p>Then the estimated Pearson correlation between the genotypes at
loci 1 and 2 is
$$\sqrt{(a1 + b1)/a1}\sqrt{(a2 + b2)/a2}r.$$
All other LD calculations are based on this equation. In particular,
the estimated genotype variances at loci 1 and 2 are
\(a1 + b1\) and \(a2 + b2\), respectively, which can be
used to calculate D and D'.</p>
<p>The reliability ratio for SNP i is defined by \((ai + bi)/ai\).
By default, we apply <code><a href='https://rdrr.io/pkg/ashr/man/ash.html'>ash</a>()</code> (Stephens, 2016)
to the log of these reliability ratios before adjusting the
Pearson correlation. Standard errors are required before using
<code><a href='https://rdrr.io/pkg/ashr/man/ash.html'>ash</a>()</code>, but these are easily obtained
using the central limit theorem and the delta-method.</p>
    <h2 class="hasAnchor" id="references"><a class="anchor" href="#references"></a>References</h2>

    
<ul>
<li><p>Gerard, David. Scalable Bias-corrected Linkage Disequilibrium Estimation Under Genotype Uncertainty. <em>bioRxiv</em>, 2021. doi: <a href='https://doi.org/10.1101/2021.02.08.430270'>10.1101/2021.02.08.430270</a></p></li>
<li><p>T. Robertson and J. D. Cryer. An iterative procedure for estimating the mode. <em>Journal of the American Statistical Association</em>, 69(348):1012–1016, 1974. doi: <a href='https://doi.org/10.1080/01621459.1974.10480246'>10.1080/01621459.1974.10480246</a>
.</p></li>
<li><p>M. Stephens. False discovery rates: a new deal. <em>Biostatistics</em>, 18(2):275–294, 10 2016. ISSN 1465-4644 doi: <a href='https://doi.org/10.1093/biostatistics/kxw041'>10.1093/biostatistics/kxw041</a>
.</p></li>
</ul>

    <h2 class="hasAnchor" id="see-also"><a class="anchor" href="#see-also"></a>See also</h2>

    <div class='dont-index'>
<dl>
<dt><code><a href='gl_to_gp.html'>gl_to_gp</a>()</code></dt><dd><p>Normalize genotype likelihoods to
posterior probabilities using naive uniform prior.</p></dd>
<dt><code><a href='https://rdrr.io/pkg/ashr/man/ash.html'>ash</a>()</code></dt><dd><p>Function used to perform hierarchical
shrinkage on the log of the reliability ratios.</p></dd>
<dt><code><a href='ldest.html'>ldest</a>()</code>, <code><a href='mldest.html'>mldest</a>()</code>, <code><a href='sldest.html'>sldest</a>()</code></dt><dd><p>Maximum likelihood estimation of linkage disequilibrium.</p></dd>

</dl>
</div>
    <h2 class="hasAnchor" id="author"><a class="anchor" href="#author"></a>Author</h2>

    <p>David Gerard</p>

    <h2 class="hasAnchor" id="examples"><a class="anchor" href="#examples"></a>Examples</h2>
    <pre class="examples"><div class='input'><span class='fu'><a href='https://rdrr.io/r/utils/data.html'>data</a></span><span class='op'>(</span><span class='st'>"gp"</span><span class='op'>)</span>

<span class='va'>ldout</span> <span class='op'>&lt;-</span> <span class='fu'>ldfast</span><span class='op'>(</span><span class='va'>gp</span>, <span class='st'>"r"</span><span class='op'>)</span>
<span class='va'>ldout</span><span class='op'>$</span><span class='va'>ldmat</span>
</div><div class='output co'>#&gt;            [,1]      [,2]       [,3]         [,4]      [,5]      [,6]
#&gt;  [1,] 1.0000000 1.0000000 0.43269941  0.831256306 0.6028276 0.6028276
#&gt;  [2,] 1.0000000 1.0000000 0.43269941  0.831256306 0.6028276 0.6028276
#&gt;  [3,] 0.4326994 0.4326994 1.00000000  0.573455065 0.8927188 0.8927188
#&gt;  [4,] 0.8312563 0.8312563 0.57345506  1.000000000 0.5484621 0.5484621
#&gt;  [5,] 0.6028276 0.6028276 0.89271875  0.548462093 1.0000000 1.0000000
#&gt;  [6,] 0.6028276 0.6028276 0.89271875  0.548462093 1.0000000 1.0000000
#&gt;  [7,] 0.4281823 0.4281823 0.97896686  0.581033314 0.8885195 0.8885195
#&gt;  [8,] 0.4281823 0.4281823 0.97896686  0.581033314 0.8885195 0.8885195
#&gt;  [9,] 0.1036140 0.1036140 0.09310332 -0.001987959 0.1340953 0.1340953
#&gt; [10,] 0.5571129 0.5571129 0.82400859  0.540610084 0.9605484 0.9605484
#&gt;             [,7]       [,8]         [,9]     [,10]
#&gt;  [1,] 0.42818226 0.42818226  0.103613990 0.5571129
#&gt;  [2,] 0.42818226 0.42818226  0.103613990 0.5571129
#&gt;  [3,] 0.97896686 0.97896686  0.093103317 0.8240086
#&gt;  [4,] 0.58103331 0.58103331 -0.001987959 0.5406101
#&gt;  [5,] 0.88851952 0.88851952  0.134095304 0.9605484
#&gt;  [6,] 0.88851952 0.88851952  0.134095304 0.9605484
#&gt;  [7,] 1.00000000 1.00000000  0.079131179 0.8512614
#&gt;  [8,] 1.00000000 1.00000000  0.079131179 0.8512614
#&gt;  [9,] 0.07913118 0.07913118  1.000000000 0.1020791
#&gt; [10,] 0.85126139 0.85126139  0.102079061 1.0000000</div><div class='input'><span class='va'>ldout</span><span class='op'>$</span><span class='va'>rr</span>
</div><div class='output co'>#&gt;  [1] 1.005771 1.005771 1.003719 1.003953 1.001617 1.001617 1.003960 1.003960
#&gt;  [9] 1.062503 1.032458</div><div class='input'><span class='va'>ldout</span><span class='op'>$</span><span class='va'>semat</span>
</div><div class='output co'>#&gt;             [,1]       [,2]       [,3]       [,4]       [,5]       [,6]
#&gt;  [1,]         NA         NA 0.09332343 0.03572633 0.06723054 0.06723054
#&gt;  [2,]         NA         NA 0.09332343 0.03572633 0.06723054 0.06723054
#&gt;  [3,] 0.09332343 0.09332343         NA 0.06374431 0.02960981 0.02960981
#&gt;  [4,] 0.03572633 0.03572633 0.06374431         NA 0.06308449 0.06308449
#&gt;  [5,] 0.06723054 0.06723054 0.02960981 0.06308449         NA         NA
#&gt;  [6,] 0.06723054 0.06723054 0.02960981 0.06308449         NA         NA
#&gt;  [7,] 0.09234704 0.09234704 0.01306726 0.06062951 0.02913919 0.02913919
#&gt;  [8,] 0.09234704 0.09234704 0.01306726 0.06062951 0.02913919 0.02913919
#&gt;  [9,] 0.11330929 0.11330929 0.11338916 0.10962925 0.10304775 0.10304775
#&gt; [10,] 0.07796815 0.07796815 0.04525697 0.06934184 0.01956130 0.01956130
#&gt;             [,7]       [,8]      [,9]      [,10]
#&gt;  [1,] 0.09234704 0.09234704 0.1133093 0.07796815
#&gt;  [2,] 0.09234704 0.09234704 0.1133093 0.07796815
#&gt;  [3,] 0.01306726 0.01306726 0.1133892 0.04525697
#&gt;  [4,] 0.06062951 0.06062951 0.1096292 0.06934184
#&gt;  [5,] 0.02913919 0.02913919 0.1030478 0.01956130
#&gt;  [6,] 0.02913919 0.02913919 0.1030478 0.01956130
#&gt;  [7,]         NA         NA 0.1152616 0.04149810
#&gt;  [8,]         NA         NA 0.1152616 0.04149810
#&gt;  [9,] 0.11526155 0.11526155        NA 0.11264141
#&gt; [10,] 0.04149810 0.04149810 0.1126414         NA</div><div class='input'>
<span class='va'>ldout</span> <span class='op'>&lt;-</span> <span class='fu'>ldfast</span><span class='op'>(</span><span class='va'>gp</span>, <span class='st'>"D"</span><span class='op'>)</span>
<span class='va'>ldout</span><span class='op'>$</span><span class='va'>ldmat</span>
</div><div class='output co'>#&gt;             [,1]       [,2]       [,3]          [,4]       [,5]       [,6]
#&gt;  [1,] 0.24316990 0.24316990 0.11133968  0.1890926874 0.14190970 0.14190970
#&gt;  [2,] 0.24316990 0.24316990 0.11133968  0.1890926874 0.14190970 0.14190970
#&gt;  [3,] 0.11133968 0.11133968 0.27228111  0.1380362082 0.22237575 0.22237575
#&gt;  [4,] 0.18909269 0.18909269 0.13803621  0.2127990758 0.12078013 0.12078013
#&gt;  [5,] 0.14190970 0.14190970 0.22237575  0.1207801316 0.22789145 0.22789145
#&gt;  [6,] 0.14190970 0.14190970 0.22237575  0.1207801316 0.22789145 0.22789145
#&gt;  [7,] 0.11108594 0.11108594 0.26875234  0.1410137301 0.22315493 0.22315493
#&gt;  [8,] 0.11108594 0.11108594 0.26875234  0.1410137301 0.22315493 0.22315493
#&gt;  [9,] 0.01152076 0.01152076 0.01095422 -0.0002067759 0.01443395 0.01443395
#&gt; [10,] 0.12264546 0.12264546 0.19195255  0.1113326112 0.20470887 0.20470887
#&gt;              [,7]        [,8]          [,9]      [,10]
#&gt;  [1,] 0.111085938 0.111085938  0.0115207554 0.12264546
#&gt;  [2,] 0.111085938 0.111085938  0.0115207554 0.12264546
#&gt;  [3,] 0.268752337 0.268752337  0.0109542225 0.19195255
#&gt;  [4,] 0.141013730 0.141013730 -0.0002067759 0.11133261
#&gt;  [5,] 0.223154931 0.223154931  0.0144339511 0.20470887
#&gt;  [6,] 0.223154931 0.223154931  0.0144339511 0.20470887
#&gt;  [7,] 0.276790383 0.276790383  0.0093870857 0.19993638
#&gt;  [8,] 0.276790383 0.276790383  0.0093870857 0.19993638
#&gt;  [9,] 0.009387086 0.009387086  0.0508411413 0.01027538
#&gt; [10,] 0.199936377 0.199936377  0.0102753754 0.19929976</div><div class='input'><span class='va'>ldout</span><span class='op'>$</span><span class='va'>rr</span>
</div><div class='output co'>#&gt;  [1] 1.011575 1.011575 1.007451 1.007921 1.003237 1.003237 1.007936 1.007936
#&gt;  [9] 1.128912 1.065970</div><div class='input'><span class='va'>ldout</span><span class='op'>$</span><span class='va'>semat</span>
</div><div class='output co'>#&gt;             [,1]       [,2]       [,3]       [,4]       [,5]       [,6]
#&gt;  [1,]         NA         NA 0.03145416 0.03197325 0.03082308 0.03082308
#&gt;  [2,]         NA         NA 0.03145416 0.03197325 0.03082308 0.03082308
#&gt;  [3,] 0.03145416 0.03145416         NA 0.02713108 0.03349861 0.03349861
#&gt;  [4,] 0.03197325 0.03197325 0.02713108         NA 0.02596956 0.02596956
#&gt;  [5,] 0.03082308 0.03082308 0.03349861 0.02596956         NA         NA
#&gt;  [6,] 0.03082308 0.03082308 0.03349861 0.02596956         NA         NA
#&gt;  [7,] 0.03101405 0.03101405 0.03348837 0.02704158 0.03328525 0.03328525
#&gt;  [8,] 0.03101405 0.03101405 0.03348837 0.02704158 0.03328525 0.03328525
#&gt;  [9,] 0.01277533 0.01277533 0.01328870 0.01127557 0.01137312 0.01137312
#&gt; [10,] 0.02665070 0.02665070 0.02646479 0.02337153 0.02738681 0.02738681
#&gt;             [,7]       [,8]       [,9]      [,10]
#&gt;  [1,] 0.03101405 0.03101405 0.01277533 0.02665070
#&gt;  [2,] 0.03101405 0.03101405 0.01277533 0.02665070
#&gt;  [3,] 0.03348837 0.03348837 0.01328870 0.02646479
#&gt;  [4,] 0.02704158 0.02704158 0.01127557 0.02337153
#&gt;  [5,] 0.03328525 0.03328525 0.01137312 0.02738681
#&gt;  [6,] 0.03328525 0.03328525 0.01137312 0.02738681
#&gt;  [7,]         NA         NA 0.01360690 0.02608113
#&gt;  [8,]         NA         NA 0.01360690 0.02608113
#&gt;  [9,] 0.01360690 0.01360690         NA 0.01132382
#&gt; [10,] 0.02608113 0.02608113 0.01132382         NA</div><div class='input'>
<span class='va'>ldout</span> <span class='op'>&lt;-</span> <span class='fu'>ldfast</span><span class='op'>(</span><span class='va'>gp</span>, <span class='st'>"Dprime"</span><span class='op'>)</span>
<span class='va'>ldout</span><span class='op'>$</span><span class='va'>ldmat</span>
</div><div class='output co'>#&gt;             [,1]       [,2]       [,3]          [,4]      [,5]      [,6]
#&gt;  [1,] 4.00000000 0.99234186 0.68861855  0.9333391533 1.1182071 1.1182071
#&gt;  [2,] 0.99234186 4.00000000 0.68861855  0.9333391533 1.1182071 1.1182071
#&gt;  [3,] 0.68861855 0.68861855 4.00000000  1.0445593471 1.3592571 1.3592571
#&gt;  [4,] 0.93333915 0.93333915 1.04455935  4.0000000000 1.1644397 1.1644397
#&gt;  [5,] 1.11820712 1.11820712 1.35925711  1.1644397472 4.0000000 1.2815177
#&gt;  [6,] 1.11820712 1.11820712 1.35925711  1.1644397472 1.2815177 4.0000000
#&gt;  [7,] 0.67835286 0.67835286 1.29638385  1.0535843861 1.3714162 1.3714162
#&gt;  [8,] 0.67835286 0.67835286 1.29638385  1.0535843861 1.3714162 1.3714162
#&gt;  [9,] 0.04904818 0.04904818 0.06492008 -0.0008680408 0.1089843 0.1089843
#&gt; [10,] 0.96807877 0.96807877 1.17532263  1.0752091511 1.1494195 1.1494195
#&gt;             [,7]       [,8]          [,9]      [,10]
#&gt;  [1,] 0.67835286 0.67835286  0.0490481819 0.96807877
#&gt;  [2,] 0.67835286 0.67835286  0.0490481819 0.96807877
#&gt;  [3,] 1.29638385 1.29638385  0.0649200760 1.17532263
#&gt;  [4,] 1.05358439 1.05358439 -0.0008680408 1.07520915
#&gt;  [5,] 1.37141620 1.37141620  0.1089843435 1.14941949
#&gt;  [6,] 1.37141620 1.37141620  0.1089843435 1.14941949
#&gt;  [7,] 4.00000000 1.32871897  0.0549282885 1.23084578
#&gt;  [8,] 1.32871897 4.00000000  0.0549282885 1.23084578
#&gt;  [9,] 0.05492829 0.05492829  4.0000000000 0.07771872
#&gt; [10,] 1.23084578 1.23084578  0.0777187155 4.00000000</div><div class='input'><span class='va'>ldout</span><span class='op'>$</span><span class='va'>rr</span>
</div><div class='output co'>#&gt;  [1] 1.011575 1.011575 1.007451 1.007921 1.003237 1.003237 1.007936 1.007936
#&gt;  [9] 1.128912 1.065970</div><div class='input'><span class='va'>ldout</span><span class='op'>$</span><span class='va'>semat</span>
</div><div class='output co'>#&gt;             [,1]       [,2]       [,3]      [,4]       [,5]       [,6]
#&gt;  [1,]         NA 0.13995724 0.19150934 0.1526757 0.21646378 0.21646378
#&gt;  [2,] 0.13995724         NA 0.19150934 0.1526757 0.21646378 0.21646378
#&gt;  [3,] 0.19150934 0.19150934         NA 0.1889567 0.17151103 0.17151103
#&gt;  [4,] 0.15267573 0.15267573 0.18895674        NA 0.22732911 0.22732911
#&gt;  [5,] 0.21646378 0.21646378 0.17151103 0.2273291         NA 0.16529356
#&gt;  [6,] 0.21646378 0.21646378 0.17151103 0.2273291 0.16529356         NA
#&gt;  [7,] 0.18673492 0.18673492 0.14851597 0.1876919 0.17147607 0.17147607
#&gt;  [8,] 0.18673492 0.18673492 0.14851597 0.1876919 0.17147607 0.17147607
#&gt;  [9,] 0.05460539 0.05460539 0.07852983 0.0473413 0.08452456 0.08452456
#&gt; [10,] 0.19999687 0.19999687 0.15592059 0.2163060 0.14446347 0.14446347
#&gt;             [,7]       [,8]       [,9]      [,10]
#&gt;  [1,] 0.18673492 0.18673492 0.05460539 0.19999687
#&gt;  [2,] 0.18673492 0.18673492 0.05460539 0.19999687
#&gt;  [3,] 0.14851597 0.14851597 0.07852983 0.15592059
#&gt;  [4,] 0.18769185 0.18769185 0.04734130 0.21630603
#&gt;  [5,] 0.17147607 0.17147607 0.08452456 0.14446347
#&gt;  [6,] 0.17147607 0.17147607 0.08452456 0.14446347
#&gt;  [7,]         NA 0.14500916 0.07949034 0.15287303
#&gt;  [8,] 0.14500916         NA 0.07949034 0.15287303
#&gt;  [9,] 0.07949034 0.07949034         NA 0.08538073
#&gt; [10,] 0.15287303 0.15287303 0.08538073         NA</div><div class='input'>
</div></pre>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top">
      <h2 data-toc-skip>Contents</h2>
    </nav>
  </div>
</div>


      <footer>
      <div class="copyright">
  <p>Developed by David Gerard.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.6.1.</p>
</div>

      </footer>
   </div>

  


  </body>
</html>


