# Generated by using Rcpp::compileAttributes() -> do not edit by hand
# Generator token: 10BE3573-1514-4C36-9D1C-5A225CD40393

#' Derivative of multinomial pdf
#'
#' Gradient of multinomial pdf (NOT log-pdf) with respect to prob,
#' optionally returned on the log-scale.
#'
#' @inheritParams dmulti_double
#'
#' @author David Gerard
#'
#' @noRd
dmulti_dprob <- function(x, prob, log_p = TRUE) {
    .Call(`_ldsep_dmulti_dprob`, x, prob, log_p)
}

#' Derivative of \code{\link{probgeno}()} with respect to \code{prob}.
#'
#' @inheritParams probgeno
#'
#' @author David Gerard
#'
#' @return A vector of length 4 containing the gradient of
#'     \code{\link{probgeno}()} with respect to \code{prob}.
#'
#' @noRd
dprobgeno_dprob <- function(gA, gB, K, prob) {
    .Call(`_ldsep_dprobgeno_dprob`, gA, gB, K, prob)
}

#' Derivative of \code{\link{proballgeno}()} with respect to \code{prob}.
#'
#' @inheritParams proballgeno
#'
#' @author David Gerard
#'
#' @return A vector of length 4 containing the gradient of
#'     \code{\link{proballgeno}()} with respect to \code{prob}.
#'
#' @noRd
dproballgeno_dprob <- function(gA, gB, K, prob) {
    .Call(`_ldsep_dproballgeno_dprob`, gA, gB, K, prob)
}

#' Derivative of \code{\link{real_to_simplex}()} with respect to \code{y}.
#'
#' @param y A numeric matrix.
#'
#' @author David Gerard
#'
#' @noRd
dreal_to_simplex_dy <- function(y) {
    .Call(`_ldsep_dreal_to_simplex_dy`, y)
}

#' Derivative of \code{\link{simplex_to_real}()} with respect to \code{x}.
#'
#' @param x A simplex elements.
#'
#' @author David Gerard
#'
#' @noRd
dsimplex_to_real_dx <- function(x) {
    .Call(`_ldsep_dsimplex_to_real_dx`, x)
}

#' Derivative of \code{\link{llike_geno}()} with respect to par.
#'
#'
#' @inheritParams llike_geno
#'
#' @author David Gerard
#'
#' @noRd
dllike_geno_dpar <- function(par, gA, gB, K) {
    .Call(`_ldsep_dllike_geno_dpar`, par, gA, gB, K)
}

#' Derivative of prob[[4]] - (prob[[2]] + prob[[4]]) * (prob[[3]] + prob[[4]])
#' with respect to prob.
#'
#' These derivatives are with repsect to prob[1], prob[2], and prob[3].
#' prob[4] is defined in terms of those other three values as
#' prob[4] = 1 - sum(prob[1:3]).
#'
#' @param prob Probability vector in order of (ab, Ab, aB, AB)
#'
#' @author David Gerard
#'
#' @noRd
dD_dprob <- function(prob) {
    .Call(`_ldsep_dD_dprob`, prob)
}

#' Derivative of squared correlation with respect to prob.
#'
#' These derivatives are with repsect to prob[1], prob[2], and prob[3].
#' prob[4] is defined in terms of those other three values as
#' prob[4] = 1 - sum(prob[1:3]).
#'
#' @param prob Probability vector in order of (ab, Ab, aB, AB)
#'
#' @author David Gerard
#'
#' @noRd
dr2_dprob <- function(prob) {
    .Call(`_ldsep_dr2_dprob`, prob)
}

#' Derivative of D'with respect to prob.
#'
#' These derivatives are with repsect to prob[1], prob[2], and prob[3].
#' prob[4] is defined in terms of those other three values as
#' prob[4] = 1 - sum(prob[1:3]).
#'
#' @param prob Probability vector in order of (ab, Ab, aB, AB)
#'
#' @author David Gerard
#'
#' @noRd
dDprime_dprob <- function(prob) {
    .Call(`_ldsep_dDprime_dprob`, prob)
}

#' Multinomial pdf
#'
#' @param x A vector of counts
#' @param prob A vector of probabilities
#' @param log_p A logical. Should we return the log (\code{TRUE}) or not
#'     (\code{FALSE})?
#'
#' @author David Gerard
#'
#' @noRd
dmulti_double <- function(x, prob, log_p = TRUE) {
    .Call(`_ldsep_dmulti_double`, x, prob, log_p)
}

#' Probability of genotypes given haplotype frequencies for one individual
#'
#' @param gA The number of A alleles.
#' @param gB The number of B alleles.
#' @param K The ploidy of the species.
#' @param prob A numeric vector with the probabilities of haplotypes
#'     (in order) of (ab, Ab, aB, AB).
#' @param log_p A logical. log_p A logical. Should we return the log
#'      (\code{TRUE}) or not (\code{FALSE})?
#'
#' @author David Gerard
#'
#' @noRd
probgeno <- function(gA, gB, K, prob, log_p = TRUE) {
    .Call(`_ldsep_probgeno`, gA, gB, K, prob, log_p)
}

#' Probability of genotypes given haplotype frequencies for all individuals
#'
#' @param gA Vector of number of A alleles.
#' @param gB Vector of number of B alleles.
#' @inheritParams probgeno
#'
#' @author David Gerard
#'
#' @noRd
proballgeno <- function(gA, gB, K, prob, log_p = TRUE) {
    .Call(`_ldsep_proballgeno`, gA, gB, K, prob, log_p)
}

#' Likelihood function when estimating LD directly from genotypes
#'
#' @param par A vector of length 3 containing real numbers that are to
#'     be transformed into the simplex of prob (ab, Ab, aB, AB).
#' @inheritParams proballgeno
#'
#' @author David Gerard
#'
#' @noRd
llike_geno <- function(par, gA, gB, K) {
    .Call(`_ldsep_llike_geno`, par, gA, gB, K)
}

#' Find LD estimates using just the genotypes.
#'
#'
#' @param par The parameters on the real-scale.
#' @param gA The genotypes at locus 1.
#' @param gB The genotypes at locus 2.
#' @param K The ploidy for the species. Assumed to be the same for
#'     all individuals.
#' @param reltol The stopping criterion for the gradient ascent.
#'
#' @author David Gerard
#'
#' @noRd
optimize_genocor <- function(par, gA, gB, K, reltol = 10.0e-08) {
    .Call(`_ldsep_optimize_genocor`, par, gA, gB, K, reltol)
}

#' Log-sum-exponential trick.
#'
#' @param x A vector to log-sum-exp.
#'
#' @return The log of the sum of the exponential
#'     of the elements in \code{x}.
#'
#' @author David Gerard
#'
#' @noRd
log_sum_exp <- function(x) {
    .Call(`_ldsep_log_sum_exp`, x)
}

#' Log-sum-exponential trick using just two doubles.
#'
#' @param x A double.
#' @param y Another double.
#'
#' @return The log of the sum of the exponential of x and y.
#'
#' @author David Gerard
#'
#' @noRd
log_sum_exp_2 <- function(x, y) {
    .Call(`_ldsep_log_sum_exp_2`, x, y)
}

#' Pair-wise log-sum-exponential
#'
#' Does pair-wise log-sum-exponential on two vectors.
#'
#' @param x A numeric vector.
#' @param y Another numeric vector.
#'
#' @author David Gerard
#'
#' @noRd
plog_sum_exp <- function(x, y) {
    .Call(`_ldsep_plog_sum_exp`, x, y)
}

#' The logit function.
#'
#' @param x A double between 0 and 1.
#'
#' @return The logit of \code{x}.
#'
#' @author David Gerard
#'
#' @keywords internal
#' @noRd
#'
logit <- function(x) {
    .Call(`_ldsep_logit`, x)
}

#' The expit (logistic) function.
#'
#' @param x A double.
#'
#' @return The expit (logistic) of \code{x}.
#'
#' @keywords internal
#' @noRd
#'
#' @author David Gerard
expit <- function(x) {
    .Call(`_ldsep_expit`, x)
}

#' Convert real line to simplex using Stan technique
#'
#' @param y A vector of numbers of length K-1
#'
#' @return A vector on the simplex of length K
#'
#' @author David Gerard
#'
#' @seealso \code{\link{simplex_to_real}()} for the inverse function.
#'
#' @noRd
real_to_simplex <- function(y) {
    .Call(`_ldsep_real_to_simplex`, y)
}

#' Convert simplex to real-line using Stan technique
#'
#' @param x A vector on the simplex of length K.
#'
#' @return A vector of numbers of length K-1
#'
#' @author David Gerard
#'
#' @seealso \code{\link{real_to_simplex}()} for the inverse function.
#'
#' @noRd
simplex_to_real <- function(x) {
    .Call(`_ldsep_simplex_to_real`, x)
}

